// asyncbox/build/lib/asyncbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseCancellationError = exports.TimeoutError = void 0;
exports.sleep = sleep;
exports.withTimeout = withTimeout;
exports.longSleep = longSleep;
exports.retry = retry;
exports.retryInterval = retryInterval;
exports.asyncmap = asyncmap;
exports.asyncfilter = asyncfilter;
exports.waitForCondition = waitForCondition;
const p_limit_1 = require("p-limit");
const LONG_SLEEP_THRESHOLD = 5000; // anything over 5000ms will turn into a spin
/** Error thrown by {@link withTimeout} when the deadline is exceeded. */
class TimeoutError extends Error {
    constructor(message) {
        super(message ?? 'Operation timed out');
        this.name = 'TimeoutError';
    }
}
exports.TimeoutError = TimeoutError;
/** Thrown when a promise is cancelled via `cancel`. */
class PromiseCancellationError extends Error {
    constructor(message = 'Promise cancelled') {
        super(message);
        this.name = 'PromiseCancellationError';
    }
}
exports.PromiseCancellationError = PromiseCancellationError;
function sleep(arg) {
    const { ms, cancelError } = parseSleepArg(arg);
    let timeoutId;
    let resolveFn;
    let rejectFn;
    const promise = new Promise((resolve, reject) => {
        resolveFn = resolve;
        rejectFn = reject;
        timeoutId = setTimeout(resolve, ms);
    });
    promise.cancel = () => {
        if (timeoutId) {
            clearTimeout(timeoutId);
            timeoutId = undefined;
        }
        if (cancelError === null) {
            resolveFn?.(undefined);
        }
        else {
            let err;
            if (typeof cancelError === 'string' && cancelError) {
                err = new PromiseCancellationError(cancelError);
            }
            else if (cancelError instanceof Error) {
                err = cancelError;
            }
            else {
                err = new PromiseCancellationError();
            }
            rejectFn?.(err);
        }
        resolveFn = undefined;
        rejectFn = undefined;
    };
    return promise;
}
/**
 * Resolves with `promise` or rejects if it does not settle before `timeoutMs`.
 *
 * If the deadline passes first: with no third argument (or a falsy non-error value), rejects
 * with {@link TimeoutError} whose message includes `timeoutMs`; with a non-empty string,
 * rejects with {@link TimeoutError} using that message; with an `Error` instance, rejects
 * with that same instance.
 *
 * @param promise - Promise to race against the timeout.
 * @param timeoutMs - Maximum time in milliseconds before rejecting if `promise` has not settled.
 * @param messageOrError - Optional override for the timeout rejection: custom {@link TimeoutError}
 * message when the string is non-empty, or an existing `Error` to reject with verbatim.
 * Omitted, `undefined`, or other falsy non-error values use the default timeout message (includes `timeoutMs`).
 */
async function withTimeout(promise, timeoutMs, messageOrError) {
    let timer = null;
    try {
        return await Promise.race([
            promise,
            new Promise((_resolve, reject) => {
                timer = setTimeout(() => {
                    if (typeof messageOrError === 'string' && messageOrError) {
                        reject(new TimeoutError(messageOrError));
                    }
                    else if (!messageOrError) {
                        reject(new TimeoutError(`Operation timed out after ${timeoutMs}ms`));
                    }
                    else {
                        reject(messageOrError);
                    }
                }, timeoutMs);
            }),
        ]);
    }
    finally {
        if (timer) {
            clearTimeout(timer);
        }
    }
}
/**
 * Sometimes `Promise.delay` or `setTimeout` are inaccurate for large wait
 * times. To safely wait for these long times (e.g. in the 5+ minute range), you
 * can use `longSleep`.
 *
 * You can also pass a `progressCb` option which is a callback function that
 * receives an object with the properties `elapsedMs`, `timeLeft`, and
 * `progress`. This will be called on every wait interval so you can do your
 * wait logging or whatever.
 * @param ms - The number of milliseconds to wait
 * @param options - Options for controlling the long sleep behavior
 */
async function longSleep(ms, { thresholdMs = LONG_SLEEP_THRESHOLD, intervalMs = 1000, progressCb = null } = {}) {
    if (ms < thresholdMs) {
        return await sleep(ms);
    }
    const endAt = Date.now() + ms;
    let timeLeft;
    let elapsedMs = 0;
    do {
        const pre = Date.now();
        await sleep(intervalMs);
        const post = Date.now();
        timeLeft = endAt - post;
        elapsedMs = elapsedMs + (post - pre);
        if (typeof progressCb === 'function') {
            progressCb({ elapsedMs, timeLeft, progress: elapsedMs / ms });
        }
    } while (timeLeft > 0);
}
/**
 * An async/await way of running a method until it doesn't throw an error
 * @param times - The maximum number of times to retry the function
 * @param fn - The async function to retry
 * @param args - Arguments to pass to the function
 */
async function retry(times, fn, ...args) {
    let tries = 0;
    let done = false;
    let res = null;
    while (!done && tries < times) {
        tries++;
        try {
            res = await fn(...args);
            done = true;
        }
        catch (err) {
            if (tries >= times) {
                throw err;
            }
        }
    }
    return res;
}
/**
 * You can also use `retryInterval` to add a sleep in between retries. This can
 * be useful if you want to throttle how fast we retry.
 * @param times - The maximum number of times to retry the function
 * @param sleepMs - The number of milliseconds to wait between retries
 * @param fn - The async function to retry
 * @param args - Arguments to pass to the function
 */
async function retryInterval(times, sleepMs, fn, ...args) {
    let count = 0;
    const wrapped = async () => {
        count++;
        let res;
        try {
            res = await fn(...args);
        }
        catch (e) {
            // do not pause when finished the last retry
            if (count !== times) {
                await sleep(sleepMs);
            }
            throw e;
        }
        return res;
    };
    return await retry(times, wrapped);
}
/**
 * Similar to `Array.prototype.map`; runs in parallel, serial, or with custom concurrency pool
 * @param coll - The collection to map over
 * @param mapper - The function to apply to each element
 * @param options - Options for controlling parallelism (default: true - fully parallel)
 */
async function asyncmap(coll, mapper, options = true) {
    if (options === null) {
        throw new Error('Options cannot be null');
    }
    // limitFunction requires the mapper to always return a promise
    const mapperAsync = async (value) => mapper(value);
    if (options === false) {
        return coll.reduce(async (acc, item) => [...(await acc), await mapperAsync(item)], Promise.resolve([]));
    }
    const adjustedMapper = options === true ? mapperAsync : (0, p_limit_1.limitFunction)(mapperAsync, { concurrency: options.concurrency });
    return Promise.all(coll.map(adjustedMapper));
}
/**
 * Similar to `Array.prototype.filter`; runs in parallel, serial, or with custom concurrency pool
 * @param coll - The collection to filter
 * @param filter - The function to test each element
 * @param options - Options for controlling parallelism (default: true - fully parallel)
 */
async function asyncfilter(coll, filter, options = true) {
    if (options === null) {
        throw new Error('Options cannot be null');
    }
    // limitFunction requires the filter to always return a promise
    const filterAsync = async (value) => filter(value);
    if (options === false) {
        return coll.reduce(async (accP, item) => {
            const acc = await accP;
            if (await filterAsync(item)) {
                acc.push(item);
            }
            return acc;
        }, Promise.resolve([]));
    }
    const adjustedFilter = options === true ? filterAsync : (0, p_limit_1.limitFunction)(filterAsync, { concurrency: options.concurrency });
    const bools = await Promise.all(coll.map(adjustedFilter));
    return coll.reduce((acc, item, i) => {
        if (bools[i]) {
            acc.push(item);
        }
        return acc;
    }, []);
}
/**
 * Takes a condition (a function returning a boolean or boolean promise), and
 * waits until the condition is true.
 *
 * Throws a `/Condition unmet/` error if the condition has not been satisfied
 * within the allocated time, unless an error is provided in the options, as the
 * `error` property, which is either thrown itself, or used as the message.
 *
 * The condition result is returned if it is not falsy. If the condition throws an
 * error then this exception will be immediately passed through.
 *
 * The default options are: `{ waitMs: 5000, intervalMs: 500 }`
 * @param condFn - The condition function to evaluate
 * @param options - Options for controlling the wait behavior
 */
async function waitForCondition(condFn, options = {}) {
    const opts = {
        ...options,
        waitMs: typeof options.waitMs === 'number' ? options.waitMs : 5000,
        intervalMs: typeof options.intervalMs === 'number' ? options.intervalMs : 500,
    };
    const debug = opts.logger ? opts.logger.debug.bind(opts.logger) : () => undefined;
    const error = opts.error;
    const begunAt = Date.now();
    const endAt = begunAt + opts.waitMs;
    const spin = async function spin() {
        const result = await condFn();
        if (result) {
            return result;
        }
        const now = Date.now();
        const waited = now - begunAt;
        const remainingTime = endAt - now;
        if (now < endAt) {
            debug(`Waited for ${waited} ms so far`);
            await sleep(Math.min(opts.intervalMs, remainingTime));
            return await spin();
        }
        // if there is an error option, it is either a string message or an error itself
        if (error) {
            throw typeof error === 'string' ? new Error(error) : error;
        }
        throw new Error(`Condition unmet after ${waited} ms. Timing out.`);
    };
    return await spin();
}
/** Non-array object values (including class instances); excludes `null` and arrays. */
function isSleepArgObject(value) {
    return value !== null && typeof value === 'object' && !Array.isArray(value);
}
function parseSleepArg(arg) {
    if (typeof arg === 'number') {
        if (!Number.isFinite(arg)) {
            throw new TypeError('sleep: expected a finite number or an object with ms');
        }
        return { ms: arg };
    }
    if (isSleepArgObject(arg)) {
        const ms = arg.ms;
        if (typeof ms !== 'number' || !Number.isFinite(ms)) {
            throw new TypeError('sleep: options.ms must be a finite number');
        }
        return { ms, cancelError: arg.cancelError };
    }
    throw new TypeError('sleep: expected a finite number or an object with ms');
}
