// @appium/support/build/lib/timing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = exports.Timer = exports.Duration = void 0;
const NS_PER_S = 1e9;
const NS_PER_MS = 1e6;
/**
 * Class representing a duration, encapsulating the number and units.
 */
class Duration {
    _nanos;
    constructor(_nanos) {
        this._nanos = _nanos;
    }
    get nanos() {
        return this._nanos;
    }
    /**
     * Get the duration as nanoseconds
     *
     * @returns {number} The duration as nanoseconds
     */
    get asNanoSeconds() {
        return this.nanos;
    }
    /**
     * Get the duration converted into milliseconds
     *
     * @returns {number} The duration as milliseconds
     */
    get asMilliSeconds() {
        return this.nanos / NS_PER_MS;
    }
    /**
     * Get the duration converted into seconds
     *
     * @returns {number} The duration fas seconds
     */
    get asSeconds() {
        return this.nanos / NS_PER_S;
    }
    toString() {
        // default to milliseconds, rounded
        return this.asMilliSeconds.toFixed(0);
    }
}
exports.Duration = Duration;
class Timer {
    _startTime;
    /**
     * Creates a timer
     */
    constructor(_startTime = null) {
        this._startTime = _startTime;
    }
    get startTime() {
        return this._startTime;
    }
    /**
     * Start the timer
     *
     * @return {Timer} The current instance, for chaining
     */
    start() {
        if (this._startTime !== null) {
            throw new Error('Timer has already been started.');
        }
        this._startTime = process.hrtime.bigint();
        return this;
    }
    /**
     * Get the duration since the timer was started
     *
     * @return {Duration} the duration
     */
    getDuration() {
        if (this._startTime === null) {
            throw new Error('Unable to get duration. Timer was not started');
        }
        let nanoDuration;
        if (Array.isArray(this._startTime)) {
            // startTime was created using process.hrtime()
            const [seconds, nanos] = process.hrtime(this._startTime);
            nanoDuration = seconds * NS_PER_S + nanos;
        }
        else if (typeof this._startTime === 'bigint') {
            // startTime was created using process.hrtime.bigint()
            const endTime = process.hrtime.bigint();
            // get the difference, and convert to number
            nanoDuration = Number(endTime - this._startTime);
        }
        else {
            throw new Error(`Unable to get duration. Start time '${this._startTime}' cannot be parsed`);
        }
        return new Duration(nanoDuration);
    }
    toString() {
        try {
            return this.getDuration().toString();
        }
        catch (err) {
            return `<err: ${err.message}>`;
        }
    }
}
exports.Timer = Timer;
exports.default = Timer;
