// @appium/support/build/lib/zip.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractAllTo = extractAllTo;
exports._extractEntryTo = _extractEntryTo;
exports.readEntries = readEntries;
exports.toInMemoryZip = toInMemoryZip;
exports.assertValidZip = assertValidZip;
exports.toArchive = toArchive;
const node_util_1 = require("node:util");
const yauzl = __importStar(require("yauzl"));
const archiver_1 = __importDefault(require("archiver"));
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const node_stream_1 = __importDefault(require("node:stream"));
const promises_1 = require("node:stream/promises");
const fs_1 = require("./fs");
const system_1 = require("./system");
const base64_stream_1 = require("base64-stream");
const util_1 = require("./util");
const timing_1 = require("./timing");
const logger_1 = __importDefault(require("./logger"));
const get_stream_1 = __importDefault(require("get-stream"));
const teen_process_1 = require("teen_process");
const openZip = (0, node_util_1.promisify)(yauzl.open);
const ZIP_MAGIC = 'PK';
const IFMT = 0b1111000000000000;
const IFDIR = 0b0100000000000000;
const IFLNK = 0b1010000000000000;
// This class is mostly copied from https://github.com/maxogden/extract-zip/blob/master/index.js
class ZipExtractor {
    zipfile;
    zipPath;
    opts;
    canceled = false;
    constructor(sourcePath, opts) {
        this.zipPath = sourcePath;
        this.opts = opts;
    }
    async extract() {
        const { dir, fileNamesEncoding } = this.opts;
        this.zipfile = await openZip(this.zipPath, {
            lazyEntries: true,
            // https://github.com/thejoshwolfe/yauzl/commit/cc7455ac789ba84973184e5ebde0581cdc4c3b39#diff-04c6e90faac2675aa89e2176d2eec7d8R95
            decodeStrings: !fileNamesEncoding,
        });
        this.canceled = false;
        return new Promise((resolve, reject) => {
            this.zipfile.on('error', (err) => {
                this.canceled = true;
                reject(err);
            });
            this.zipfile.readEntry();
            this.zipfile.on('close', () => {
                if (!this.canceled) {
                    resolve();
                }
            });
            this.zipfile.on('entry', async (entry) => {
                if (this.canceled) {
                    return;
                }
                const fileName = this.extractFileName(entry);
                if (fileName.startsWith('__MACOSX/')) {
                    this.zipfile.readEntry();
                    return;
                }
                const destDir = node_path_1.default.dirname(node_path_1.default.join(dir, fileName));
                try {
                    await fs_1.fs.mkdir(destDir, { recursive: true });
                    const canonicalDestDir = await fs_1.fs.realpath(destDir);
                    const relativeDestDir = node_path_1.default.relative(dir, canonicalDestDir);
                    if (relativeDestDir.split(node_path_1.default.sep).includes('..')) {
                        throw new Error(`Out of bound path "${canonicalDestDir}" found while processing file ${fileName}`);
                    }
                    await this.extractEntry(entry);
                    this.zipfile.readEntry();
                }
                catch (err) {
                    this.canceled = true;
                    this.zipfile.close();
                    reject(err);
                }
            });
        });
    }
    extractFileName(entry) {
        if (Buffer.isBuffer(entry.fileName)) {
            return entry.fileName.toString(this.opts.fileNamesEncoding);
        }
        return entry.fileName;
    }
    async extractEntry(entry) {
        if (this.canceled) {
            return;
        }
        const { dir } = this.opts;
        const fileName = this.extractFileName(entry);
        const dest = node_path_1.default.join(dir, fileName);
        // convert external file attr int into a fs stat mode int
        const mode = (entry.externalFileAttributes >> 16) & 0xffff;
        // check if it's a symlink or dir (using stat mode constants)
        const isSymlink = (mode & IFMT) === IFLNK;
        const isDir = (mode & IFMT) === IFDIR ||
            // Failsafe, borrowed from jsZip
            fileName.endsWith('/') ||
            // check for windows weird way of specifying a directory
            // https://github.com/maxogden/extract-zip/issues/13#issuecomment-154494566
            (entry.versionMadeBy >> 8 === 0 && entry.externalFileAttributes === 16);
        const procMode = this.getExtractedMode(mode, isDir) & 0o777;
        // always ensure folders are created
        const destDir = isDir ? dest : node_path_1.default.dirname(dest);
        const mkdirOptions = { recursive: true };
        if (isDir) {
            mkdirOptions.mode = procMode;
        }
        await fs_1.fs.mkdir(destDir, mkdirOptions);
        if (isDir) {
            return;
        }
        const openReadStream = (0, node_util_1.promisify)(this.zipfile.openReadStream.bind(this.zipfile));
        const readStream = await openReadStream(entry);
        if (isSymlink) {
            const link = await (0, get_stream_1.default)(readStream);
            await fs_1.fs.symlink(link, dest);
        }
        else {
            await (0, promises_1.pipeline)(readStream, fs_1.fs.createWriteStream(dest, { mode: procMode }));
        }
    }
    getExtractedMode(entryMode, isDir) {
        const { defaultDirMode, defaultFileMode } = this.opts;
        let mode = entryMode;
        // Set defaults, if necessary
        if (mode === 0) {
            if (isDir) {
                if (defaultDirMode) {
                    mode = parseInt(defaultDirMode, 10);
                }
                if (!mode) {
                    mode = 0o755;
                }
            }
            else {
                if (defaultFileMode) {
                    mode = parseInt(defaultFileMode, 10);
                }
                if (!mode) {
                    mode = 0o644;
                }
            }
        }
        return mode;
    }
}
/**
 * Extract zipfile to a directory
 *
 * @param zipFilePath The full path to the source ZIP file
 * @param destDir The full path to the destination folder
 * @param opts Extraction options
 */
async function extractAllTo(zipFilePath, destDir, opts = {}) {
    if (!node_path_1.default.isAbsolute(destDir)) {
        throw new Error(`Target path '${destDir}' is expected to be absolute`);
    }
    await fs_1.fs.mkdir(destDir, { recursive: true });
    const dir = await fs_1.fs.realpath(destDir);
    if (opts.useSystemUnzip) {
        try {
            await extractWithSystemUnzip(zipFilePath, dir);
            return;
        }
        catch (err) {
            logger_1.default.warn('unzip failed; falling back to JS: %s', err.stderr || err.message);
        }
    }
    const extractor = new ZipExtractor(zipFilePath, {
        ...opts,
        dir,
    });
    await extractor.extract();
}
/**
 * Extract a single zip entry to a directory
 *
 * @private
 * @param zipFile The source ZIP stream
 * @param entry The entry instance
 * @param destDir The full path to the destination folder
 */
async function _extractEntryTo(zipFile, entry, destDir) {
    const dstPath = node_path_1.default.resolve(destDir, entry.fileName);
    if (!(0, util_1.isSubPath)(dstPath, destDir)) {
        throw new Error(`Out of bound path "${dstPath}" found while processing file ${entry.fileName}`);
    }
    // Create dest directory if doesn't exist already
    if (entry.fileName.endsWith('/')) {
        if (!(await fs_1.fs.exists(dstPath))) {
            await fs_1.fs.mkdirp(dstPath);
        }
        return;
    }
    else if (!(await fs_1.fs.exists(node_path_1.default.dirname(dstPath)))) {
        await fs_1.fs.mkdirp(node_path_1.default.dirname(dstPath));
    }
    // Create a write stream
    const writeStream = (0, node_fs_1.createWriteStream)(dstPath, { flags: 'w' });
    const writeStreamPromise = new Promise((resolve, reject) => {
        writeStream.once('finish', resolve);
        writeStream.once('error', reject);
    });
    const openReadStream = (0, node_util_1.promisify)(zipFile.openReadStream.bind(zipFile));
    const zipReadStream = await openReadStream(entry);
    const zipReadStreamPromise = new Promise((resolve, reject) => {
        zipReadStream.once('end', resolve);
        zipReadStream.once('error', reject);
    });
    zipReadStream.pipe(writeStream);
    // Wait for the zipReadStream and writeStream to end before returning
    await Promise.all([zipReadStreamPromise, writeStreamPromise]);
}
/**
 * Get entries for a zip folder
 *
 * @param zipFilePath The full path to the source ZIP file
 * @param onEntry Callback when entry is read.
 * The callback is expected to accept one argument of ZipEntry type.
 * The iteration through the source zip file will be terminated as soon as
 * the result of this function equals to `false`.
 */
async function readEntries(zipFilePath, onEntry) {
    // Open a zip file and start reading entries
    const zipfile = await openZip(zipFilePath, { lazyEntries: true });
    const zipReadStreamPromise = new Promise((resolve, reject) => {
        zipfile.once('end', resolve);
        zipfile.once('error', reject);
        // On each entry, call 'onEntry' and then read the next entry
        zipfile.on('entry', async (entry) => {
            const res = await onEntry({
                entry,
                extractEntryTo: async (destDir) => await _extractEntryTo(zipfile, entry, destDir),
            });
            if (res === false) {
                return zipfile.emit('end');
            }
            zipfile.readEntry();
        });
    });
    zipfile.readEntry();
    // Wait for the entries to finish being iterated through
    await zipReadStreamPromise;
}
/**
 * Converts contents of local directory to an in-memory .zip buffer
 *
 * @param srcPath The full path to the folder or file being zipped
 * @param opts Zipping options
 * @returns Zipped (and encoded if `encodeToBase64` is truthy)
 * content of the source path as memory buffer
 * @throws {Error} if there was an error while reading the source
 * or the source is too big
 */
async function toInMemoryZip(srcPath, opts = {}) {
    if (!(await fs_1.fs.exists(srcPath))) {
        throw new Error(`No such file or folder: ${srcPath}`);
    }
    const { isMetered = true, encodeToBase64 = false, maxSize = 1 * util_1.GiB, level = 9 } = opts;
    const resultBuffers = [];
    let resultBuffersSize = 0;
    // Create a writable stream that zip buffers will be streamed to
    const resultWriteStream = new node_stream_1.default.Writable({
        write(buffer, _encoding, next) {
            resultBuffers.push(buffer);
            resultBuffersSize += buffer.length;
            if (maxSize > 0 && resultBuffersSize > maxSize) {
                resultWriteStream.emit('error', new Error(`The size of the resulting ` +
                    `archive must not be greater than ${(0, util_1.toReadableSizeString)(maxSize)}`));
            }
            next();
        },
    });
    // Zip 'srcDir' and stream it to the above writable stream
    const archive = (0, archiver_1.default)('zip', {
        zlib: { level },
    });
    let srcSize = null;
    const base64EncoderStream = encodeToBase64 ? new base64_stream_1.Base64Encode() : null;
    const resultWriteStreamPromise = new Promise((resolve, reject) => {
        resultWriteStream.once('error', (e) => {
            if (base64EncoderStream) {
                archive.unpipe(base64EncoderStream);
                base64EncoderStream.unpipe(resultWriteStream);
            }
            else {
                archive.unpipe(resultWriteStream);
            }
            archive.abort();
            archive.destroy();
            reject(e);
        });
        resultWriteStream.once('finish', () => {
            srcSize = archive.pointer();
            resolve();
        });
    });
    const archiveStreamPromise = new Promise((resolve, reject) => {
        archive.once('finish', resolve);
        archive.once('error', (e) => reject(new Error(`Failed to archive '${srcPath}': ${e.message}`)));
    });
    const timer = isMetered ? new timing_1.Timer().start() : null;
    if ((await fs_1.fs.stat(srcPath)).isDirectory()) {
        archive.directory(srcPath, false);
    }
    else {
        archive.file(srcPath, {
            name: node_path_1.default.basename(srcPath),
        });
    }
    if (base64EncoderStream) {
        archive.pipe(base64EncoderStream);
        base64EncoderStream.pipe(resultWriteStream);
    }
    else {
        archive.pipe(resultWriteStream);
    }
    await Promise.all([archive.finalize(), archiveStreamPromise, resultWriteStreamPromise]);
    if (timer) {
        logger_1.default.debug(`Zipped ${encodeToBase64 ? 'and base64-encoded ' : ''}` +
            `'${node_path_1.default.basename(srcPath)}' ` +
            (srcSize ? `(${(0, util_1.toReadableSizeString)(srcSize)}) ` : '') +
            `in ${timer.getDuration().asSeconds.toFixed(3)}s ` +
            `(compression level: ${level})`);
    }
    // Return the array of zip buffers concatenated into one buffer
    return Buffer.concat(resultBuffers);
}
/**
 * Verifies whether the given file is a valid ZIP archive
 *
 * @param filePath - Full path to the file
 * @throws {Error} If the file does not exist or is not a valid ZIP archive
 */
async function assertValidZip(filePath) {
    if (!(await fs_1.fs.exists(filePath))) {
        throw new Error(`The file at '${filePath}' does not exist`);
    }
    const { size } = await fs_1.fs.stat(filePath);
    if (size < 4) {
        throw new Error(`The file at '${filePath}' is too small to be a ZIP archive`);
    }
    const fd = await fs_1.fs.open(filePath, 'r');
    try {
        const buffer = Buffer.alloc(ZIP_MAGIC.length);
        await fs_1.fs.read(fd, buffer, 0, ZIP_MAGIC.length, 0);
        const signature = buffer.toString('ascii');
        if (signature !== ZIP_MAGIC) {
            throw new Error(`The file signature '${signature}' of '${filePath}' ` +
                `is not equal to the expected ZIP archive signature '${ZIP_MAGIC}'`);
        }
        return true;
    }
    finally {
        await fs_1.fs.close(fd);
    }
}
/**
 * Creates an archive based on the given glob pattern
 *
 * @param dstPath - The resulting archive path
 * @param src - Source options
 * @param opts - Compression options
 * @throws {Error} If there was an error while creating the archive
 */
async function toArchive(dstPath, src = {}, opts = {}) {
    const { level = 9 } = opts;
    const { pattern = '**/*', cwd = node_path_1.default.dirname(dstPath), ignore = [] } = src;
    const archive = (0, archiver_1.default)('zip', { zlib: { level } });
    const outStream = fs_1.fs.createWriteStream(dstPath);
    await new Promise((resolve, reject) => {
        const outFinished = new Promise((_resolve, _reject) => {
            outStream.once('error', (e) => {
                archive.unpipe(outStream);
                archive.abort();
                archive.destroy();
                _reject(e);
            });
            outStream.once('finish', () => _resolve());
        });
        archive
            .glob(pattern, {
            cwd,
            ignore,
        })
            .on('error', reject)
            .pipe(outStream);
        void Promise.all([archive.finalize(), outFinished])
            .then(() => resolve())
            .catch(reject);
    });
}
/**
 * Executes system unzip (e.g., `/usr/bin/unzip`). If available, it is
 * significantly faster than the JS implementation.
 * By default all files in the destDir get overridden if already exist.
 *
 * @param zipFilePath The full path to the source ZIP file
 * @param destDir The full path to the destination folder.
 * This folder is expected to already exist before extracting the archive.
 */
async function extractWithSystemUnzip(zipFilePath, destDir) {
    const isWindowsHost = (0, system_1.isWindows)();
    let executablePath;
    try {
        executablePath = await getExecutablePath(isWindowsHost ? 'powershell.exe' : 'unzip');
    }
    catch (e) {
        throw new Error('Could not find system unzip', { cause: e });
    }
    if (isWindowsHost) {
        // on Windows we use PowerShell to unzip files
        await (0, teen_process_1.exec)(executablePath, [
            '-command',
            'Expand-Archive',
            '-LiteralPath',
            zipFilePath,
            '-DestinationPath',
            destDir,
            '-Force',
        ]);
    }
    else {
        // -q means quiet (no stdout)
        // -o means overwrite
        // -d is the dest dir
        await (0, teen_process_1.exec)(executablePath, ['-q', '-o', zipFilePath, '-d', destDir]);
    }
}
/**
 * Finds and memoizes the full path to the given executable.
 * Rejects if it is not found.
 */
const getExecutablePath = (0, util_1.memoize)(
/**
 * @returns Full Path to the executable
 */
async function getExecutablePath(binaryName) {
    const fullPath = await fs_1.fs.which(binaryName);
    logger_1.default.debug(`Found '${binaryName}' at '${fullPath}'`);
    return fullPath;
});
exports.default = {
    extractAllTo,
    readEntries,
    toInMemoryZip,
    assertValidZip,
    toArchive,
};
