// @wdio/image-comparison-core/dist/commands/saveWebElement.js
import { takeElementScreenshot } from '../methods/takeElementScreenshots.js';
import beforeScreenshot from '../helpers/beforeScreenshot.js';
import afterScreenshot from '../helpers/afterScreenshot.js';
import { DEFAULT_RESIZE_DIMENSIONS } from '../helpers/constants.js';
import { canUseBidiScreenshot, getMethodOrWicOption } from '../helpers/utils.js';
import { createBeforeScreenshotOptions, buildAfterScreenshotOptions } from '../helpers/options.js';
import { determineWebElementIgnoreRegions } from '../methods/rectangles.js';
/**
 * Saves an image of an element
 */
export default async function saveWebElement({ browserInstance, instanceData, folders, element, tag, ignore, saveElementOptions, }) {
    // 1. Set some variables
    const { addressBarShadowPadding, autoElementScroll } = saveElementOptions.wic;
    const enableLegacyScreenshotMethod = getMethodOrWicOption(saveElementOptions.method, saveElementOptions.wic, 'enableLegacyScreenshotMethod');
    const resizeDimensions = saveElementOptions.method.resizeDimensions || DEFAULT_RESIZE_DIMENSIONS;
    // 2.  Prepare the screenshot
    const beforeOptions = createBeforeScreenshotOptions(instanceData, saveElementOptions.method, saveElementOptions.wic);
    const enrichedInstanceData = await beforeScreenshot(browserInstance, beforeOptions, true);
    const { deviceName, dimensions: { window: { devicePixelRatio, innerHeight, isEmulated, isLandscape, }, }, initialDevicePixelRatio, isAndroid, isAndroidChromeDriverScreenshot, isAndroidNativeWebScreenshot, isIOS, isMobile, } = enrichedInstanceData;
    // 3. Take the screenshot
    const elementScreenshotOptions = {
        addressBarShadowPadding,
        autoElementScroll,
        deviceName,
        devicePixelRatio: devicePixelRatio || 1,
        deviceRectangles: instanceData.deviceRectangles,
        element,
        isEmulated,
        initialDevicePixelRatio: initialDevicePixelRatio || 1,
        innerHeight,
        isAndroidNativeWebScreenshot,
        isAndroidChromeDriverScreenshot,
        isAndroid,
        isIOS,
        isLandscape,
        isMobile,
        resizeDimensions,
        toolBarShadowPadding: beforeOptions.toolBarShadowPadding,
    };
    const shouldUseBidi = canUseBidiScreenshot(browserInstance) && !isMobile && !enableLegacyScreenshotMethod;
    const screenshotData = await takeElementScreenshot(browserInstance, elementScreenshotOptions, shouldUseBidi);
    // 3b. Resolve ignore regions (element-local) while the DOM is still in screenshot state.
    // determineWebElementIgnoreRegions returns device-pixel regions, or CSS-pixel regions when
    // the element image is from the native driver on Android native web (see that function).
    const ignoreRegionPadding = getMethodOrWicOption(saveElementOptions.method, saveElementOptions.wic, 'ignoreRegionPadding') ?? 1;
    const ignoreRegions = ignore && ignore.length > 0
        ? await (async () => {
            const rootElement = await element;
            return determineWebElementIgnoreRegions({
                browserInstance,
                devicePixelRatio: devicePixelRatio || 1,
                rootElement: rootElement,
                ignoreRegionPadding,
                isAndroidNativeWebScreenshot,
                isWebDriverElementScreenshot: screenshotData.isWebDriverElementScreenshot,
            }, ignore);
        })()
        : undefined;
    // 4. Return the data
    const afterOptions = buildAfterScreenshotOptions({
        base64Image: screenshotData.base64Image,
        folders,
        tag,
        isNativeContext: false,
        instanceData,
        enrichedInstanceData,
        beforeOptions,
        wicOptions: saveElementOptions.wic
    });
    const result = await afterScreenshot(browserInstance, afterOptions);
    return {
        ...result,
        ...(ignoreRegions ? { ignoreRegions } : {}),
    };
}
