// appium/build/lib/cli/driver-command.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = __importDefault(require("lodash"));
const extension_command_1 = __importDefault(require("./extension-command"));
const constants_1 = require("../constants");
require("@colors/colors");
const REQ_DRIVER_FIELDS = ['driverName', 'automationName', 'platformNames', 'mainClass'];
class DriverCliCommand extends extension_command_1.default {
    constructor({ config, json }) {
        super({ config, json });
        this.knownExtensions = constants_1.KNOWN_DRIVERS;
    }
    /**
     * Install a driver
     *
     * @param opts - install options
     */
    async install({ driver, installType, packageName }) {
        return await super._install({
            installSpec: driver,
            installType,
            packageName,
        });
    }
    /**
     * Uninstall a driver
     *
     * @param opts - uninstall options
     */
    async uninstall({ driver }) {
        return await super._uninstall({ installSpec: driver });
    }
    /**
     * Update a driver
     *
     * @param opts - update options
     */
    async update({ driver, unsafe }) {
        return await super._update({ installSpec: driver, unsafe });
    }
    /**
     * Run a script from a driver
     *
     * @param opts - script execution options
     * @throws {Error} if the script fails to run
     */
    async run({ driver, scriptName, extraArgs }) {
        return await super._run({
            installSpec: driver,
            scriptName,
            extraArgs,
            bufferOutput: this.isJsonOutput,
        });
    }
    /**
     * Runs doctor checks for the given driver.
     *
     * @param opts - doctor command options
     * @returns The amount of executed doctor checks.
     * @throws {Error} If any of the mandatory Doctor checks fails.
     */
    async doctor({ driver }) {
        return await super._doctor({
            installSpec: driver,
        });
    }
    /**
     * Builds the success message displayed after a driver installation.
     *
     * @param args - installed extension name and metadata
     * @returns formatted success text
     */
    getPostInstallText({ extName, extData }) {
        return (`Driver ${extName}@${extData.version} successfully installed\n`.green +
            `- automationName: ${extData.automationName.green}\n` +
            `- platformNames: ${JSON.stringify(extData.platformNames).green}`);
    }
    /**
     * Validates fields in `appium` field of `driverMetadata`
     *
     * For any `package.json` fields which a driver requires, validate the type of
     * those fields on the `package.json` data, throwing an error if anything is
     * amiss.
     * @param driverMetadata - `appium` metadata from extension package
     * @param installSpec - install spec from CLI
     */
    validateExtensionFields(driverMetadata, installSpec) {
        const missingFields = REQ_DRIVER_FIELDS.reduce((acc, field) => (driverMetadata[field] ? acc : [...acc, field]), []);
        if (!lodash_1.default.isEmpty(missingFields)) {
            throw new Error(`Driver "${installSpec}" did not expose correct fields for compatibility ` +
                `with Appium. Missing fields: ${JSON.stringify(missingFields)}`);
        }
    }
}
exports.default = DriverCliCommand;
