// appium/build/lib/cli/plugin-command.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = __importDefault(require("lodash"));
const extension_command_1 = __importDefault(require("./extension-command"));
const constants_1 = require("../constants");
const REQ_PLUGIN_FIELDS = ['pluginName', 'mainClass'];
class PluginCliCommand extends extension_command_1.default {
    constructor({ config, json }) {
        super({ config, json });
        this.knownExtensions = constants_1.KNOWN_PLUGINS;
    }
    /**
     * Install a plugin
     *
     * @param opts - install options
     */
    async install({ plugin, installType, packageName }) {
        return await super._install({
            installSpec: plugin,
            installType,
            packageName,
        });
    }
    /**
     * Uninstall a plugin
     *
     * @param opts - uninstall options
     */
    async uninstall({ plugin }) {
        return await super._uninstall({ installSpec: plugin });
    }
    /**
     * Update a plugin
     *
     * @param opts - update options
     */
    async update({ plugin, unsafe }) {
        return await super._update({ installSpec: plugin, unsafe });
    }
    /**
     * Run a script from a plugin
     *
     * @param opts - script execution options
     * @throws {Error} if the script fails to run
     */
    async run({ plugin, scriptName, extraArgs }) {
        return await super._run({
            installSpec: plugin,
            scriptName,
            extraArgs,
            bufferOutput: this.isJsonOutput,
        });
    }
    /**
     * Runs doctor checks for the given plugin
     *
     * @param opts - doctor command options
     * @returns The amount of executed doctor checks.
     * @throws {Error} If any of the mandatory Doctor checks fails.
     */
    async doctor({ plugin }) {
        return await super._doctor({
            installSpec: plugin,
        });
    }
    /**
     * Builds the success message displayed after a plugin installation.
     *
     * @param args - installed extension name and metadata
     * @returns formatted success text
     */
    getPostInstallText({ extName, extData }) {
        return `Plugin ${extName}@${extData.version} successfully installed`.green;
    }
    /**
     * Validates fields in `appium` field of `pluginMetadata`
     *
     * For any `package.json` fields which a plugin requires, validate the type of
     * those fields on the `package.json` data, throwing an error if anything is
     * amiss.
     * @param pluginMetadata - `appium` metadata from extension package
     * @param installSpec - install spec from CLI
     */
    validateExtensionFields(pluginMetadata, installSpec) {
        const missingFields = REQ_PLUGIN_FIELDS.reduce((acc, field) => (pluginMetadata[field] ? acc : [...acc, field]), []);
        if (!lodash_1.default.isEmpty(missingFields)) {
            throw new Error(`Installed plugin "${installSpec}" did not expose correct fields for compatibility ` +
                `with Appium. Missing fields: ${JSON.stringify(missingFields)}`);
        }
    }
}
exports.default = PluginCliCommand;
