// appium/build/lib/cli/setup-command.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_PLUGINS = exports.SUBCOMMAND_RESET = exports.SUBCOMMAND_BROWSER = exports.SUBCOMMAND_DESKTOP = exports.SUBCOMMAND_MOBILE = void 0;
exports.getPresetDrivers = getPresetDrivers;
exports.determinePlatformName = determinePlatformName;
exports.runSetupCommand = runSetupCommand;
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const extension_1 = require("./extension");
const support_1 = require("@appium/support");
const logger_1 = require("../logger");
/**
 * Subcommands of preset for setup
 */
exports.SUBCOMMAND_MOBILE = 'mobile';
exports.SUBCOMMAND_DESKTOP = 'desktop';
exports.SUBCOMMAND_BROWSER = 'browser';
exports.SUBCOMMAND_RESET = 'reset';
/**
 * Pairs of preset subcommand and driver candidates.
 * Driver names listed in KNOWN_DRIVERS to install by default
 */
const PRESET_PAIRS = Object.freeze({
    mobile: lodash_1.default.keys(constants_1.MOBILE_DRIVERS),
    desktop: lodash_1.default.keys(constants_1.DESKTOP_DRIVERS),
    browser: lodash_1.default.keys(constants_1.DESKTOP_BROWSERS),
});
const DRIVERS_ONLY_MACOS = ['xcuitest', 'safari', 'mac2'];
const DRIVERS_ONLY_WINDOWS = ['windows'];
/**
 * Plugin names listed in KNOWN_PLUGINS to install by default.
 */
exports.DEFAULT_PLUGINS = ['images', 'inspector'];
/**
 * Return a list of drivers available for current host platform.
 */
function getPresetDrivers(presetName) {
    return lodash_1.default.filter(PRESET_PAIRS[presetName], (driver) => {
        if (lodash_1.default.includes(DRIVERS_ONLY_MACOS, driver)) {
            return support_1.system.isMac();
        }
        if (lodash_1.default.includes(DRIVERS_ONLY_WINDOWS, driver)) {
            return support_1.system.isWindows();
        }
        return true;
    });
}
/**
 * Return desktop platform name for setup command description.
 */
function determinePlatformName() {
    if (support_1.system.isMac()) {
        return 'macOS';
    }
    else if (support_1.system.isWindows()) {
        return 'Windows';
    }
    return 'Linux';
}
/**
 * Runs the `setup` command and applies the selected preset.
 *
 * Depending on the subcommand, this installs mobile/desktop/browser presets or
 * removes all installed extensions and manifests via `reset`.
 */
async function runSetupCommand(preConfigArgs, driverConfig, pluginConfig) {
    switch (preConfigArgs.setupCommand) {
        case exports.SUBCOMMAND_DESKTOP:
            await setupDesktopAppDrivers(driverConfig);
            await setupDefaultPlugins(pluginConfig);
            break;
        case exports.SUBCOMMAND_BROWSER:
            await setupBrowserDrivers(driverConfig);
            await setupDefaultPlugins(pluginConfig);
            break;
        case exports.SUBCOMMAND_RESET:
            await resetAllExtensions(driverConfig, pluginConfig);
            break;
        default:
            await setupMobileDrivers(driverConfig);
            await setupDefaultPlugins(pluginConfig);
            break;
    }
}
;
/**
 * Resets all installed drivers and extensions
 */
async function resetAllExtensions(driverConfig, pluginConfig) {
    const commandConfigs = [
        ['driver', driverConfig],
        ['plugin', pluginConfig],
    ];
    for (const [command, config] of commandConfigs) {
        for (const extensionName of lodash_1.default.keys(config.installedExtensions)) {
            try {
                await uninstallExtension(extensionName, extensionCommandArgs(command, extensionName, 'uninstall'), config);
            }
            catch (e) {
                logger_1.log.warn(`${extensionName} ${command} cannot be uninstalled. Will delete the manifest anyway. ` +
                    `Original error: ${e.stack}`);
            }
        }
        const manifestPath = config.manifestPath;
        if (!manifestPath || !await support_1.fs.exists(manifestPath)) {
            continue;
        }
        await support_1.fs.rimraf(manifestPath);
        if (await support_1.fs.exists(manifestPath)) {
            throw new Error(`${command} manifest at '${manifestPath}' cannot be deleted. Is it accessible?`);
        }
        else {
            logger_1.log.info(`Successfully deleted ${command} manifest at '${manifestPath}'`);
        }
    }
}
/**
 * Install drivers listed in DEFAULT_DRIVERS.
 */
async function setupMobileDrivers(driverConfig) {
    await installDrivers(exports.SUBCOMMAND_MOBILE, driverConfig);
}
/**
 * Install all of known drivers listed in BROWSER_DRIVERS.
 */
async function setupBrowserDrivers(driverConfig) {
    await installDrivers(exports.SUBCOMMAND_BROWSER, driverConfig);
}
/**
 * Install all of known drivers listed in DESKTOP_APP_DRIVERS.
 */
async function setupDesktopAppDrivers(driverConfig) {
    await installDrivers(exports.SUBCOMMAND_DESKTOP, driverConfig);
}
/**
 * Install the given driver name. It skips the installation if the given driver name was already installed.
 */
async function installDrivers(subcommand, driverConfig) {
    for (const driverName of getPresetDrivers(subcommand)) {
        await installExtension(driverName, extensionCommandArgs('driver', driverName, 'install'), driverConfig);
    }
}
/**
 * Install plugins listed in DEFAULT_PLUGINS.
 */
async function setupDefaultPlugins(pluginConfig) {
    for (const pluginName of exports.DEFAULT_PLUGINS) {
        await installExtension(pluginName, extensionCommandArgs('plugin', pluginName, 'install'), pluginConfig);
    }
}
/**
 * Run the given extensionConfigArgs command after checking if the given extensionName was already installed.
 */
async function installExtension(extensionName, extensionConfigArgs, extensionConfig) {
    if (lodash_1.default.keys(extensionConfig.installedExtensions).includes(extensionName)) {
        logger_1.log.info(`${extensionName} (${extensionConfig.installedExtensions[extensionName].version}) is already installed. ` +
            `Skipping the installation.`);
        return;
    }
    await (0, extension_1.runExtensionCommand)(extensionConfigArgs, extensionConfig);
}
/**
 * Run the given extensionConfigArgs command after checking if the given extensionName was already installed.
 */
async function uninstallExtension(extensionName, extensionConfigArgs, extensionConfig) {
    if (!lodash_1.default.keys(extensionConfig.installedExtensions).includes(extensionName)) {
        logger_1.log.info(`${extensionName} (${extensionConfig.installedExtensions[extensionName].version}) is not installed. ` +
            `Skipping its uninstall.`);
        return;
    }
    await (0, extension_1.runExtensionCommand)(extensionConfigArgs, extensionConfig);
}
/**
 * Return the command config for driver or plugin.
 */
function extensionCommandArgs(extensionCommand, extensionName, command) {
    return (extensionCommand === 'plugin')
        ? { 'subcommand': 'plugin', 'pluginCommand': command, 'plugin': extensionName }
        : { 'subcommand': 'driver', 'driverCommand': command, 'driver': extensionName };
}
