// appium/build/lib/schema/cli-args.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toParserArgs = toParserArgs;
const argparse_1 = require("argparse");
const lodash_1 = __importDefault(require("lodash"));
const format_errors_1 = require("./format-errors");
const schema_1 = require("./schema");
const cli_transformers_1 = require("./cli-transformers");
const TYPENAMES = Object.freeze({
    ARRAY: 'array',
    OBJECT: 'object',
    BOOLEAN: 'boolean',
    INTEGER: 'integer',
    NUMBER: 'number',
    NULL: 'null',
    STRING: 'string',
});
const SHORT_ARG_CUTOFF = 3;
/**
 * Converts the finalized, flattened schema representation into
 * `ArgumentDefinitions` for handoff to `argparse`.
 */
function toParserArgs() {
    const flattened = (0, schema_1.flattenSchema)().filter(({ schema }) => !schema.appiumCliIgnored);
    return new Map(lodash_1.default.map(flattened, ({ schema, argSpec }) => subSchemaToArgDef(schema, argSpec)));
}
/**
 * Convert an alias (`foo`) to a flag (`--foo`) or short flag (`-f`).
 */
function aliasToFlag(argSpec, alias) {
    const { extType, extName, name } = argSpec;
    const arg = alias ?? name;
    const isShort = arg.length < SHORT_ARG_CUTOFF;
    if (extType && extName) {
        return isShort
            ? `--${extType}-${lodash_1.default.kebabCase(extName)}-${arg}`
            : `--${extType}-${lodash_1.default.kebabCase(extName)}-${lodash_1.default.kebabCase(arg)}`;
    }
    return isShort ? `-${arg}` : `--${lodash_1.default.kebabCase(arg)}`;
}
const screamingSnakeCase = lodash_1.default.flow(lodash_1.default.snakeCase, lodash_1.default.toUpper);
/**
 * Given an argument spec, return a validator/coercer function backed by schema validation.
 */
function getSchemaValidator({ ref: schemaId }, coerce = lodash_1.default.identity) {
    return (value) => {
        const coerced = coerce(value);
        const errors = (0, schema_1.validate)(coerced, schemaId);
        if (lodash_1.default.isEmpty(errors)) {
            return coerced;
        }
        throw new argparse_1.ArgumentTypeError('\n\n' + (0, format_errors_1.formatErrors)(errors, value, { schemaId }));
    };
}
function makeDescription(schema) {
    const { appiumCliDescription, description = '', appiumDeprecated } = schema;
    let desc = appiumCliDescription ?? description;
    if (appiumDeprecated) {
        desc = `[DEPRECATED] ${desc}`;
    }
    return desc;
}
function subSchemaToArgDef(subSchema, argSpec) {
    const { type, appiumCliAliases, appiumCliTransformer, enum: enumValues } = subSchema;
    const { name, arg } = argSpec;
    const aliases = [
        aliasToFlag(argSpec),
        ...(appiumCliAliases ?? []).map((alias) => aliasToFlag(argSpec, alias)),
    ];
    const argOpts = {
        required: false,
        help: makeDescription(subSchema),
    };
    if (!argSpec.extType) {
        argOpts.dest = argSpec.rawDest;
    }
    let argTypeFunction;
    switch (type) {
        case TYPENAMES.BOOLEAN: {
            argOpts.action = 'store_const';
            argOpts.const = true;
            break;
        }
        case TYPENAMES.OBJECT: {
            argTypeFunction = lodash_1.default.flow(cli_transformers_1.transformers.json, (o) => {
                if (!lodash_1.default.isPlainObject(o)) {
                    throw new argparse_1.ArgumentTypeError(`'${lodash_1.default.truncate(String(o), { length: 100 })}' must be a plain object`);
                }
                return o;
            });
            break;
        }
        case TYPENAMES.ARRAY: {
            argTypeFunction = cli_transformers_1.parseCsvLine;
            break;
        }
        case TYPENAMES.NUMBER: {
            argTypeFunction = getSchemaValidator(argSpec, parseFloat);
            break;
        }
        case TYPENAMES.INTEGER: {
            argTypeFunction = getSchemaValidator(argSpec, lodash_1.default.parseInt);
            break;
        }
        case TYPENAMES.STRING: {
            argTypeFunction = getSchemaValidator(argSpec);
            break;
        }
        case TYPENAMES.NULL:
        default: {
            throw new TypeError(`Schema property "${arg}": \`${type}\` type unknown or disallowed`);
        }
    }
    if (type !== TYPENAMES.BOOLEAN) {
        argOpts.metavar = screamingSnakeCase(name);
    }
    if (appiumCliTransformer && cli_transformers_1.transformers[appiumCliTransformer]) {
        if (type === TYPENAMES.ARRAY) {
            const csvTransformer = argTypeFunction;
            argTypeFunction = (val) => lodash_1.default.flatMap(csvTransformer(val).map(cli_transformers_1.transformers[appiumCliTransformer]));
        }
        else {
            argTypeFunction = lodash_1.default.flow(argTypeFunction ?? lodash_1.default.identity, cli_transformers_1.transformers[appiumCliTransformer]);
        }
    }
    if (argTypeFunction) {
        argOpts.type = argTypeFunction;
    }
    if (enumValues && !lodash_1.default.isEmpty(enumValues)) {
        if (type === TYPENAMES.STRING) {
            argOpts.choices = enumValues.map(String);
        }
        else {
            throw new TypeError(`Problem with schema for ${arg}; \`enum\` is only supported for \`type: 'string'\``);
        }
    }
    const finalAliases = aliases;
    return [finalAliases, argOpts];
}
