// appium/build/lib/schema/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAllowedSchemaFileExtension = exports.getDefaultsForExtension = exports.getDefaultsForSchema = exports.flattenSchema = exports.getSchema = exports.validate = exports.resetSchema = exports.finalizeSchema = exports.isFinalized = exports.hasArgSpec = exports.getArgSpec = exports.getAllArgSpecs = exports.registerSchema = exports.SchemaUnsupportedSchemaError = exports.SchemaUnknownSchemaError = exports.SchemaNameConflictError = exports.SchemaFinalizationError = exports.ALLOWED_SCHEMA_EXTENSIONS = exports.RoachHotelMap = void 0;
const ajv_1 = __importDefault(require("ajv"));
const ajv_formats_1 = __importDefault(require("ajv-formats"));
const lodash_1 = __importDefault(require("lodash"));
const node_path_1 = __importDefault(require("node:path"));
const schema_1 = require("@appium/schema");
const constants_1 = require("../constants");
const arg_spec_1 = require("./arg-spec");
const keywords_1 = require("./keywords");
/**
 * Key/value pairs go in... but they don't come out.
 */
class RoachHotelMap extends Map {
    set(key, value) {
        if (this.has(key)) {
            throw new Error(`${String(key)} is already set`);
        }
        return super.set(key, value);
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    delete(_key) {
        return false;
    }
    clear() {
        throw new Error(`Cannot clear RoachHotelMap`);
    }
}
exports.RoachHotelMap = RoachHotelMap;
exports.ALLOWED_SCHEMA_EXTENSIONS = Object.freeze(new Set(['.json', '.js', '.cjs']));
const SCHEMA_KEY = '$schema';
class AppiumSchema {
    static #instance;
    #argSpecs = new RoachHotelMap();
    #registeredSchemas = {
        [constants_1.DRIVER_TYPE]: new Map(),
        [constants_1.PLUGIN_TYPE]: new Map(),
    };
    #ajv;
    #finalizedSchemas = null;
    constructor() {
        this.#ajv = AppiumSchema._instantiateAjv();
    }
    /**
     * Returns a singleton instance.
     */
    static create() {
        if (!AppiumSchema.#instance) {
            const instance = new AppiumSchema();
            AppiumSchema.#instance = instance;
            lodash_1.default.bindAll(instance, [
                'finalize',
                'flatten',
                'getAllArgSpecs',
                'getArgSpec',
                'getDefaults',
                'getDefaultsForExtension',
                'getSchema',
                'hasArgSpec',
                'isFinalized',
                'registerSchema',
                'hasRegisteredSchema',
                'reset',
                'validate',
            ]);
        }
        return AppiumSchema.#instance;
    }
    /**
     * Returns `true` if filename extension is an allowed schema extension.
     */
    static isAllowedSchemaFileExtension(filename) {
        return exports.ALLOWED_SCHEMA_EXTENSIONS.has(node_path_1.default.extname(filename));
    }
    /**
     * Returns `true` if schema is a plain object and not async.
     */
    static isSupportedSchemaType(schema) {
        return lodash_1.default.isPlainObject(schema) && schema.$async !== true;
    }
    /**
     * Configures and creates an Ajv instance.
     */
    static _instantiateAjv() {
        const ajv = (0, ajv_formats_1.default)(new ajv_1.default({
            // without this not much validation actually happens
            allErrors: true,
        }));
        lodash_1.default.forEach(keywords_1.keywords, (keyword) => {
            ajv.addKeyword(keyword);
        });
        return ajv;
    }
    /**
     * Returns `true` if a schema has been registered for extension type/name.
     */
    hasRegisteredSchema(extType, extName) {
        return this.#registeredSchemas[extType].has(extName);
    }
    /**
     * Returns `true` if this instance has been finalized.
     */
    isFinalized() {
        return Boolean(this.#finalizedSchemas);
    }
    /**
     * Returns map of known argument specs.
     */
    getAllArgSpecs() {
        return this.#argSpecs;
    }
    /**
     * Finalizes all registered schemas into Ajv and generates arg-spec lookups.
     */
    async finalize() {
        if (this.isFinalized()) {
            return this.#finalizedSchemas;
        }
        const ajv = this.#ajv;
        const baseSchema = lodash_1.default.cloneDeep(schema_1.AppiumConfigJsonSchema);
        const addArgSpecs = (schema, extType, extName) => {
            for (const [propName, propSchema] of Object.entries(schema)) {
                const argSpec = arg_spec_1.ArgSpec.create(propName, {
                    dest: propSchema.appiumCliDest,
                    defaultValue: propSchema.default,
                    extType,
                    extName,
                });
                this.#argSpecs.set(argSpec.arg, argSpec);
            }
        };
        addArgSpecs(lodash_1.default.omit(baseSchema.properties.server.properties, [constants_1.DRIVER_TYPE, constants_1.PLUGIN_TYPE]));
        const finalizedSchemas = {};
        for (const [extType, extensionSchemas] of Object.entries(this.#registeredSchemas)) {
            for (const [extName, schema] of extensionSchemas.entries()) {
                const $ref = arg_spec_1.ArgSpec.toSchemaBaseRef(extType, extName);
                schema.$id = $ref;
                schema.additionalProperties = false;
                baseSchema.properties.server.properties[extType].properties[extName] = { $ref, $comment: extName };
                await ajv.validateSchema(schema, true);
                addArgSpecs(schema.properties, extType, extName);
                ajv.addSchema(schema, $ref);
                finalizedSchemas[$ref] = schema;
            }
        }
        const finalSchema = baseSchema;
        ajv.addSchema(finalSchema, arg_spec_1.APPIUM_CONFIG_SCHEMA_ID);
        finalizedSchemas[arg_spec_1.APPIUM_CONFIG_SCHEMA_ID] = finalSchema;
        await ajv.validateSchema(finalSchema, true);
        this.#finalizedSchemas = finalizedSchemas;
        return Object.freeze(finalizedSchemas);
    }
    /**
     * Resets this instance to initial state.
     */
    reset() {
        for (const schemaId of Object.keys(this.#finalizedSchemas ?? {})) {
            this.#ajv.removeSchema(schemaId);
        }
        this.#argSpecs = new RoachHotelMap();
        this.#registeredSchemas = { [constants_1.DRIVER_TYPE]: new Map(), [constants_1.PLUGIN_TYPE]: new Map() };
        this.#finalizedSchemas = null;
        this.#ajv = AppiumSchema._instantiateAjv();
    }
    /**
     * Registers an extension schema.
     */
    async registerSchema(extType, extName, schema) {
        if (!(extType && extName) || lodash_1.default.isUndefined(schema)) {
            throw new TypeError('Expected extension type, extension name, and a defined schema');
        }
        if (!AppiumSchema.isSupportedSchemaType(schema)) {
            throw new SchemaUnsupportedSchemaError(schema, extType, extName);
        }
        const normalizedExtName = lodash_1.default.kebabCase(extName);
        if (this.hasRegisteredSchema(extType, normalizedExtName)) {
            if (lodash_1.default.isEqual(this.#registeredSchemas[extType].get(normalizedExtName), schema)) {
                return;
            }
            throw new SchemaNameConflictError(extType, extName);
        }
        await this.#ajv.validateSchema(schema, true);
        this.#registeredSchemas[extType].set(normalizedExtName, schema);
    }
    /**
     * Returns an `ArgSpec` for argument name/ext context, if present.
     */
    getArgSpec(name, extType, extName) {
        return this.#argSpecs.get(arg_spec_1.ArgSpec.toArg(name, extType, extName));
    }
    /**
     * Returns `true` if an `ArgSpec` exists for argument name/ext context.
     */
    hasArgSpec(name, extType, extName) {
        return this.#argSpecs.has(arg_spec_1.ArgSpec.toArg(name, extType, extName));
    }
    /**
     * Returns default values for all args, flattened or nested.
     */
    getDefaults(flatten = true) {
        if (!this.isFinalized()) {
            throw new SchemaFinalizationError();
        }
        const reducer = flatten
            ? (defaults, { defaultValue, dest }) => {
                if (!lodash_1.default.isUndefined(defaultValue)) {
                    defaults[dest] = defaultValue;
                }
                return defaults;
            }
            : (defaults, { defaultValue, dest }) => {
                if (!lodash_1.default.isUndefined(defaultValue)) {
                    lodash_1.default.set(defaults, dest, defaultValue);
                }
                return defaults;
            };
        const retval = {};
        return [...this.#argSpecs.values()].reduce(reducer, retval);
    }
    /**
     * Returns flattened defaults for a specific extension.
     */
    getDefaultsForExtension(extType, extName) {
        if (!this.isFinalized()) {
            throw new SchemaFinalizationError();
        }
        const specs = [...this.#argSpecs.values()].filter((spec) => spec.extType === extType && spec.extName === extName);
        return specs.reduce((defaults, { defaultValue, rawDest }) => {
            if (!lodash_1.default.isUndefined(defaultValue)) {
                defaults[rawDest] = defaultValue;
            }
            return defaults;
        }, {});
    }
    /**
     * Flattens finalized schemas into schema + argSpec tuples.
     */
    flatten() {
        const schema = this.getSchema();
        const stack = [
            { properties: schema.properties, prefix: [] },
        ];
        const flattened = [];
        for (const { properties, prefix } of stack) {
            const pairs = lodash_1.default.toPairs(properties);
            for (const [key, value] of pairs) {
                if (key === SCHEMA_KEY) {
                    continue;
                }
                const { properties, $ref } = value;
                if (properties) {
                    stack.push({ properties, prefix: key === arg_spec_1.SERVER_PROP_NAME ? [] : [...prefix, key] });
                }
                else if ($ref) {
                    let refSchema;
                    try {
                        refSchema = this.getSchema($ref);
                    }
                    catch {
                        throw new SchemaUnknownSchemaError($ref);
                    }
                    const { normalizedExtName } = arg_spec_1.ArgSpec.extensionInfoFromRootSchemaId($ref);
                    if (!normalizedExtName) {
                        throw new ReferenceError(`Could not determine extension name from schema ID ${$ref}. This is a bug.`);
                    }
                    stack.push({ properties: refSchema.properties, prefix: [...prefix, key, normalizedExtName] });
                }
                else if (key !== constants_1.DRIVER_TYPE && key !== constants_1.PLUGIN_TYPE) {
                    const [extType, extName] = prefix;
                    const argSpec = this.getArgSpec(key, extType, extName);
                    if (!argSpec) {
                        throw new ReferenceError(`Unknown argument with key ${key}, extType ${extType} and extName ${extName}. This is a bug.`);
                    }
                    flattened.push({ schema: lodash_1.default.cloneDeep(value), argSpec });
                }
            }
        }
        return flattened;
    }
    /**
     * Returns schema by ID (default: base Appium schema).
     */
    getSchema(ref = arg_spec_1.APPIUM_CONFIG_SCHEMA_ID) {
        return this._getValidator(ref).schema;
    }
    /**
     * Validates a value against schema by ID/reference.
     */
    validate(value, ref = arg_spec_1.APPIUM_CONFIG_SCHEMA_ID) {
        const validator = this._getValidator(ref);
        return !validator(value) && lodash_1.default.isArray(validator.errors) ? [...validator.errors] : [];
    }
    /**
     * Retrieves schema validator function from Ajv.
     */
    _getValidator(id = arg_spec_1.APPIUM_CONFIG_SCHEMA_ID) {
        const validator = this.#ajv.getSchema(id);
        if (!validator) {
            if (id === arg_spec_1.APPIUM_CONFIG_SCHEMA_ID) {
                throw new SchemaFinalizationError();
            }
            throw new SchemaUnknownSchemaError(id);
        }
        return validator;
    }
}
class SchemaFinalizationError extends Error {
    code = 'APPIUMERR_SCHEMA_FINALIZATION';
    constructor() {
        super('Schema not yet finalized; `finalize()` must be called first.');
    }
}
exports.SchemaFinalizationError = SchemaFinalizationError;
/**
 * Thrown when a unique extension schema name conflicts with an existing one.
 */
class SchemaNameConflictError extends Error {
    code = 'APPIUMERR_SCHEMA_NAME_CONFLICT';
    data;
    constructor(extType, extName) {
        super(`Name for ${extType} schema "${extName}" conflicts with an existing schema`);
        this.data = { extType, extName };
    }
}
exports.SchemaNameConflictError = SchemaNameConflictError;
/**
 * Thrown when requested schema ID is unknown to Ajv.
 */
class SchemaUnknownSchemaError extends ReferenceError {
    code = 'APPIUMERR_SCHEMA_UNKNOWN_SCHEMA';
    data;
    constructor(schemaId) {
        super(`Unknown schema: "${schemaId}"`);
        this.data = { schemaId };
    }
}
exports.SchemaUnknownSchemaError = SchemaUnknownSchemaError;
/**
 * Thrown when provided schema type is unsupported (boolean/async/non-object).
 */
class SchemaUnsupportedSchemaError extends TypeError {
    code = 'APPIUMERR_SCHEMA_UNSUPPORTED_SCHEMA';
    data;
    constructor(schema, extType, extName) {
        super((() => {
            const msg = `Unsupported schema from ${extType} "${extName}":`;
            if (lodash_1.default.isBoolean(schema)) {
                return `${msg} schema cannot be a boolean`;
            }
            if (lodash_1.default.isPlainObject(schema)) {
                if (schema.$async) {
                    return `${msg} schema cannot be an async schema`;
                }
                throw new TypeError(`schema IS supported; this error should not be thrown (this is a bug). value of schema: ${JSON.stringify(schema)}`);
            }
            return `${msg} schema must be a plain object without a true "$async" property`;
        })());
        this.data = { schema, extType, extName };
    }
}
exports.SchemaUnsupportedSchemaError = SchemaUnsupportedSchemaError;
const appiumSchema = AppiumSchema.create();
exports.registerSchema = appiumSchema.registerSchema, exports.getAllArgSpecs = appiumSchema.getAllArgSpecs, exports.getArgSpec = appiumSchema.getArgSpec, exports.hasArgSpec = appiumSchema.hasArgSpec, exports.isFinalized = appiumSchema.isFinalized, exports.finalizeSchema = appiumSchema.finalize, exports.resetSchema = appiumSchema.reset, exports.validate = appiumSchema.validate, exports.getSchema = appiumSchema.getSchema, exports.flattenSchema = appiumSchema.flatten, exports.getDefaultsForSchema = appiumSchema.getDefaults, exports.getDefaultsForExtension = appiumSchema.getDefaultsForExtension;
exports.isAllowedSchemaFileExtension = AppiumSchema.isAllowedSchemaFileExtension;
