// appium/build/lib/extension/driver-config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DriverConfig = void 0;
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const logger_1 = require("../logger");
const extension_config_1 = require("./extension-config");
class DriverConfig extends extension_config_1.ExtensionConfig {
    static _instances = new WeakMap();
    knownAutomationNames = new Set();
    constructor(manifest) {
        super(constants_1.DRIVER_TYPE, manifest);
    }
    static create(manifest) {
        const instance = new DriverConfig(manifest);
        if (DriverConfig.getInstance(manifest)) {
            throw new Error(`Manifest with APPIUM_HOME ${manifest.appiumHome} already has a DriverConfig; use DriverConfig.getInstance() to retrieve it.`);
        }
        DriverConfig._instances.set(manifest, instance);
        return instance;
    }
    static getInstance(manifest) {
        return DriverConfig._instances.get(manifest);
    }
    async validate() {
        this.knownAutomationNames.clear();
        return await super._validate(this.manifest.getExtensionData(constants_1.DRIVER_TYPE));
    }
    extensionDesc(driverName, { version, automationName }) {
        return `${String(driverName)}@${version} (automationName '${automationName}')`;
    }
    async findMatchingDriver({ automationName, platformName, }) {
        if (!lodash_1.default.isString(platformName)) {
            throw new Error('You must include a platformName capability');
        }
        if (!lodash_1.default.isString(automationName)) {
            throw new Error('You must include an automationName capability');
        }
        logger_1.log.info(`Attempting to find matching driver for automationName ` +
            `'${automationName}' and platformName '${platformName}'`);
        try {
            const { driverName, mainClass, version } = this._getDriverBySupport(automationName, platformName);
            logger_1.log.info(`The '${driverName}' driver was installed and matched caps.`);
            logger_1.log.info(`Will require it at ${this.getInstallPath(driverName)}`);
            const driver = await this.requireAsync(driverName);
            if (!driver) {
                throw new Error(`Driver '${driverName}' did not export a class with name '${mainClass}'. Contact the author of the driver!`);
            }
            return { driver, version, driverName };
        }
        catch (err) {
            const msg = `Could not find a driver for automationName ` +
                `'${automationName}' and platformName '${platformName}'. ` +
                `Have you installed a driver that supports those ` +
                `capabilities? Run 'appium driver list --installed' to see. ` +
                `(Lower-level error: ${err.message})`;
            throw new Error(msg, { cause: err });
        }
    }
    getConfigProblems(extManifest, extName) {
        void extName;
        const problems = [];
        const { platformNames, automationName } = extManifest;
        if (!lodash_1.default.isArray(platformNames)) {
            problems.push({
                err: 'Missing or incorrect supported platformNames list.',
                val: platformNames,
            });
        }
        else if (lodash_1.default.isEmpty(platformNames)) {
            problems.push({
                err: 'Empty platformNames list.',
                val: platformNames,
            });
        }
        else {
            for (const pName of platformNames) {
                if (!lodash_1.default.isString(pName)) {
                    problems.push({
                        err: 'Incorrectly formatted platformName.',
                        val: pName,
                    });
                }
            }
        }
        if (!lodash_1.default.isString(automationName)) {
            problems.push({
                err: 'Missing or incorrect automationName',
                val: automationName,
            });
        }
        if (this.knownAutomationNames.has(automationName)) {
            problems.push({
                err: 'Multiple drivers claim support for the same automationName',
                val: automationName,
            });
        }
        this.knownAutomationNames.add(automationName);
        return problems;
    }
    _getDriverBySupport(matchAutomationName, matchPlatformName) {
        const drivers = this.installedExtensions;
        for (const [driverName, driverData] of lodash_1.default.toPairs(drivers)) {
            const { automationName, platformNames } = driverData;
            const aNameMatches = automationName.toLowerCase() === matchAutomationName.toLowerCase();
            const pNameMatches = lodash_1.default.includes(platformNames.map((p) => lodash_1.default.toLower(p)), matchPlatformName.toLowerCase());
            if (aNameMatches && pNameMatches) {
                return { driverName, ...driverData };
            }
            if (aNameMatches) {
                throw new Error(`Driver '${driverName}' supports automationName ` +
                    `'${automationName}', but Appium could not find ` +
                    `support for platformName '${matchPlatformName}'. Supported ` +
                    `platformNames are: ` +
                    JSON.stringify(platformNames));
            }
        }
        throw new Error(`Could not find installed driver to support given caps`);
    }
}
exports.DriverConfig = DriverConfig;
