// appium/build/lib/extension/extension-config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionConfig = exports.INSTALL_TYPES = exports.INSTALL_TYPE_DEV = exports.INSTALL_TYPE_GIT = exports.INSTALL_TYPE_GITHUB = exports.INSTALL_TYPE_LOCAL = exports.INSTALL_TYPE_NPM = void 0;
exports.resolveEsmEntryPoint = resolveEsmEntryPoint;
const support_1 = require("@appium/support");
const bluebird_1 = __importDefault(require("bluebird"));
const lodash_1 = __importDefault(require("lodash"));
const node_path_1 = __importDefault(require("node:path"));
const node_url_1 = require("node:url");
const resolve_from_1 = __importDefault(require("resolve-from"));
const semver_1 = require("semver");
const extension_1 = require("../cli/extension");
const build_1 = require("../helpers/build");
const logger_1 = require("../logger");
const schema_1 = require("../schema/schema");
const DEFAULT_ENTRY_POINT = 'index.js';
/**
 * "npm" install type
 * Used when extension was installed by npm package name
 * @remarks _All_ extensions are installed _by_ `npm`, but only this one means the package name was
 * used to specify it
 */
exports.INSTALL_TYPE_NPM = 'npm';
/**
 * "local" install type
 * Used when extension was installed from a local path
 */
exports.INSTALL_TYPE_LOCAL = 'local';
/**
 * "github" install type
 * Used when extension was installed via GitHub URL
 */
exports.INSTALL_TYPE_GITHUB = 'github';
/**
 * "git" install type
 * Used when extensions was installed via Git URL
 */
exports.INSTALL_TYPE_GIT = 'git';
/**
 * "dev" install type
 * Used when automatically detected as a working copy
 */
exports.INSTALL_TYPE_DEV = 'dev';
exports.INSTALL_TYPES = new Set([
    exports.INSTALL_TYPE_GIT,
    exports.INSTALL_TYPE_GITHUB,
    exports.INSTALL_TYPE_LOCAL,
    exports.INSTALL_TYPE_NPM,
    exports.INSTALL_TYPE_DEV,
]);
/**
 * Shared configuration and validation for installed Appium extensions (drivers or plugins).
 * Subclasses fix the extension kind; do not instantiate this class directly.
 */
class ExtensionConfig {
    extensionType;
    manifest;
    installedExtensions;
    #listDataCache;
    constructor(extensionType, manifest) {
        this.extensionType = extensionType;
        this.manifest = manifest;
        this.installedExtensions = manifest.getExtensionData(extensionType);
    }
    /** Path to `extensions.yaml` after the manifest has been read; otherwise undefined. */
    get manifestPath() {
        return this.manifest.manifestPath;
    }
    /** `APPIUM_HOME` directory this config is tied to. */
    get appiumHome() {
        return this.manifest.appiumHome;
    }
    /**
     * Type guard: manifest entry includes a `schema` path or inline schema object.
     *
     * @param extManifest - Parsed extension metadata
     */
    static extDataHasSchema(extManifest) {
        return lodash_1.default.isString(extManifest?.schema) || lodash_1.default.isObject(extManifest?.schema);
    }
    static async _readExtensionSchema(appiumHome, extType, extName, extManifest) {
        const { pkgName, schema: argSchemaPath } = extManifest;
        if (!argSchemaPath) {
            throw new TypeError(`No \`schema\` property found in config for ${extType} ${pkgName} -- why is this function being called?`);
        }
        let moduleObject;
        if (lodash_1.default.isString(argSchemaPath)) {
            const schemaPath = (0, resolve_from_1.default)(appiumHome, node_path_1.default.join(pkgName, argSchemaPath));
            moduleObject = require(schemaPath);
        }
        else {
            moduleObject = argSchemaPath;
        }
        // this sucks. default exports should be destroyed
        const schema = moduleObject.__esModule ? moduleObject.default : moduleObject;
        await (0, schema_1.registerSchema)(extType, extName, schema);
        return schema;
    }
    /**
     * Collects blocking validation issues for one extension (generic fields, type-specific rules, and schema).
     *
     * @param extName - Extension key as stored in the manifest
     * @param extManifest - Manifest entry for that extension
     */
    async getProblems(extName, extManifest) {
        return [
            ...this.getGenericConfigProblems(extManifest, extName),
            ...this.getConfigProblems(extManifest, extName),
            ...(await this.getSchemaProblems(extManifest, extName)),
        ];
    }
    /**
     * Collects non-fatal issues for one extension (e.g. manifest quirks, peer dependency mismatches).
     * Warnings do not by themselves prevent loading.
     *
     * @param extName - Extension key as stored in the manifest
     * @param extManifest - Manifest entry for that extension
     */
    async getWarnings(extName, extManifest) {
        const [genericConfigWarnings, configWarnings] = await bluebird_1.default.all([
            this.getGenericConfigWarnings(extManifest, extName),
            this.getConfigWarnings(extManifest, extName),
        ]);
        return [...genericConfigWarnings, ...configWarnings];
    }
    /**
     * Turns per-extension errors and warnings into human-readable log lines for console output.
     *
     * @param errorMap - Extension name to list of blocking problems
     * @param warningMap - Extension name to list of warning strings
     */
    getValidationResultSummaries(errorMap = new Map(), warningMap = new Map()) {
        const errorSummaries = [];
        for (const [extName, problems] of errorMap.entries()) {
            if (lodash_1.default.isEmpty(problems)) {
                continue;
            }
            // remove this extension from the list since it's not valid
            errorSummaries.push(`${this.extensionType} "${extName}" had ${support_1.util.pluralize('error', problems.length)} and will not be available:`);
            for (const problem of problems) {
                errorSummaries.push(`  - ${problem.err} (Actual value: ` + `${JSON.stringify(problem.val)})`);
            }
        }
        const warningSummaries = [];
        for (const [extName, warnings] of warningMap.entries()) {
            if (lodash_1.default.isEmpty(warnings)) {
                continue;
            }
            const extTypeText = lodash_1.default.capitalize(this.extensionType);
            const problemEnumerationText = support_1.util.pluralize('potential problem', warnings.length, true);
            warningSummaries.push(`${extTypeText} "${extName}" has ${problemEnumerationText}: `);
            for (const warning of warnings) {
                warningSummaries.push(`  - ${warning}`);
            }
        }
        return { errorSummaries, warningSummaries };
    }
    /**
     * Records a new installed extension in the manifest and optionally persists immediately.
     *
     * @param extName - Manifest key for the extension
     * @param extManifest - Full manifest payload
     *
     * Pass `{ write: false }` to defer flushing until a later manifest write.
     */
    async addExtension(extName, extManifest, { write = true } = {}) {
        this.manifest.setExtension(this.extensionType, extName, extManifest);
        if (write) {
            await this.manifest.write();
        }
    }
    /**
     * Merges new metadata into an existing extension entry and optionally writes the manifest.
     *
     * @param extName - Installed extension to update
     * @param extManifest - Fields to merge over the current entry
     *
     * Pass `{ write: false }` to defer flushing until a later manifest write.
     */
    async updateExtension(extName, extManifest, { write = true } = {}) {
        const existing = this.installedExtensions[extName];
        this.manifest.setExtension(this.extensionType, extName, {
            ...existing,
            ...extManifest,
        });
        if (write) {
            await this.manifest.write();
        }
    }
    /**
     * Drops an extension from the manifest and optionally persists immediately.
     *
     * @param extName - Installed extension to remove
     *
     * Pass `{ write: false }` to defer flushing until a later manifest write.
     */
    async removeExtension(extName, { write = true } = {}) {
        this.manifest.deleteExtension(this.extensionType, extName);
        if (write) {
            await this.manifest.write();
        }
    }
    /**
     * Logs installed extensions to the console. Subclasses may use `activeNames` to annotate active plugins.
     */
    print(_activeNames) {
        void _activeNames;
        if (lodash_1.default.isEmpty(this.installedExtensions)) {
            logger_1.log.info(`No ${this.extensionType}s have been installed in ${this.appiumHome}. Use the "appium ${this.extensionType}" ` +
                'command to install the one(s) you want to use.');
            return;
        }
        logger_1.log.info(`Available ${this.extensionType}s:`);
        for (const [extName, extManifest] of lodash_1.default.toPairs(this.installedExtensions)) {
            logger_1.log.info(`  - ${this.extensionDesc(extName, extManifest)}`);
        }
    }
    /**
     * Root directory of an installed extension, preferring `installPath` and falling back to `node_modules/<pkgName>`.
     *
     * @param extName - Installed extension key
     */
    getInstallPath(extName) {
        return (this.installedExtensions[extName]?.installPath ??
            node_path_1.default.join(this.appiumHome, 'node_modules', this.installedExtensions[extName].pkgName));
    }
    /**
     * Dynamically imports the extension entry point and returns the exported main class constructor.
     *
     * @param extName - Installed extension to load
     */
    async requireAsync(extName) {
        const [reqPath, mainClass] = await this._resolveExtension(extName);
        logger_1.log.debug(`Requiring ${this.extensionType} at ${reqPath}`);
        // https://github.com/nodejs/node/issues/31710
        const importPath = support_1.system.isWindows() ? (0, node_url_1.pathToFileURL)(reqPath).href : reqPath;
        const mod = (await import(importPath));
        const MainClass = mod[mainClass];
        if (!MainClass) {
            throw new ReferenceError(`Could not find a class named "${mainClass}" exported by ${this.extensionType} "${extName}"`);
        }
        return MainClass;
    }
    /** Whether the manifest lists an extension under the given name. */
    isInstalled(extName) {
        return extName in this.installedExtensions;
    }
    /**
     * Loads the extension’s config schema from disk or inline JSON and registers it for CLI/config validation.
     *
     * @param extName - Extension key
     * @param extManifest - Manifest entry that includes `schema`
     */
    async readExtensionSchema(extName, extManifest) {
        return await ExtensionConfig._readExtensionSchema(this.appiumHome, this.extensionType, extName, extManifest);
    }
    /** Optional async warnings for this extension kind; override in subclasses when needed. */
    async getConfigWarnings(_extManifest, _extName) {
        void _extManifest;
        void _extName;
        return [];
    }
    /**
     * Validates all entries in `exts`, logs summaries, and removes keys that have blocking errors.
     * Intended for subclasses’ `validate` implementation.
     */
    async _validate(exts) {
        const errorMap = new Map();
        const warningMap = new Map();
        for (const [extName, extManifest] of lodash_1.default.toPairs(exts)) {
            const [errors, warnings] = await bluebird_1.default.all([
                this.getProblems(extName, extManifest),
                this.getWarnings(extName, extManifest),
            ]);
            if (errors.length) {
                delete exts[extName];
            }
            errorMap.set(extName, errors);
            warningMap.set(extName, warnings);
        }
        const { errorSummaries, warningSummaries } = this.getValidationResultSummaries(errorMap, warningMap);
        if (!lodash_1.default.isEmpty(errorSummaries)) {
            logger_1.log.error(`Appium encountered ${support_1.util.pluralize('error', errorMap.size, true)} while validating ${this.extensionType}s found in manifest ${this.manifestPath}`);
            for (const summary of errorSummaries) {
                logger_1.log.error(summary);
            }
        }
        else if (!lodash_1.default.isEmpty(warningSummaries)) {
            // only display warnings if there are no errors!
            logger_1.log.warn(`Appium encountered ${support_1.util.pluralize('warning', warningMap.size, true)} while validating ${this.extensionType}s found in manifest ${this.manifestPath}`);
            for (const summary of warningSummaries) {
                logger_1.log.warn(summary);
            }
        }
        return exts;
    }
    /**
     * Fetches `appium driver|plugin list`-style data via the CLI command class; result is cached.
     */
    async getListData() {
        if (this.#listDataCache) {
            return this.#listDataCache;
        }
        const CommandClass = extension_1.commandClasses[this.extensionType];
        const cmd = new CommandClass({ config: this, json: true });
        const listData = await cmd.list({ showInstalled: true, showUpdates: true });
        this.#listDataCache = listData;
        return listData;
    }
    /**
     * Warnings about manifest install fields and Appium peer dependency compatibility for one extension.
     */
    async getGenericConfigWarnings(extManifest, extName) {
        const { appiumVersion, installSpec, installType, pkgName } = extManifest;
        const warnings = [];
        const invalidFields = [];
        if (!lodash_1.default.isString(installSpec)) {
            invalidFields.push('installSpec');
        }
        if (!exports.INSTALL_TYPES.has(installType)) {
            invalidFields.push('installType');
        }
        const extTypeText = lodash_1.default.capitalize(this.extensionType);
        if (invalidFields.length) {
            const invalidFieldsEnumerationText = support_1.util.pluralize('invalid or missing field', invalidFields.length, true);
            const invalidFieldsText = invalidFields.map((field) => `"${field}"`).join(', ');
            warnings.push(`${extTypeText} "${extName}" (package \`${pkgName}\`) has ${invalidFieldsEnumerationText} (${invalidFieldsText}) in \`extensions.yaml\`; ` +
                `this may cause upgrades done via the \`appium\` CLI tool to fail. Please reinstall with \`appium ${this.extensionType} uninstall ` +
                `${extName}\` and \`appium ${this.extensionType} install ${extName}\` to attempt a fix.`);
        }
        const createPeerWarning = (reason) => `${extTypeText} "${extName}" (package \`${pkgName}\`) may be incompatible with the current version of Appium (v${build_1.APPIUM_VER}) due to ${reason}`;
        if (lodash_1.default.isString(appiumVersion) && !(0, semver_1.satisfies)(build_1.APPIUM_VER, appiumVersion)) {
            const listData = await this.getListData();
            const extListData = listData[extName];
            if (extListData?.installed) {
                const { updateVersion, upToDate } = extListData;
                if (!upToDate && updateVersion) {
                    warnings.push(createPeerWarning(`its peer dependency on Appium ${appiumVersion}. Try to upgrade \`${pkgName}\` to v${updateVersion} or newer.`));
                }
                else {
                    warnings.push(createPeerWarning(`its peer dependency on Appium ${appiumVersion}. Please install a compatible version of the ${lodash_1.default.toLower(extTypeText)}.`));
                }
            }
        }
        else if (!lodash_1.default.isString(appiumVersion)) {
            const listData = await this.getListData();
            const extListData = listData[extName];
            if (!extListData?.upToDate && extListData?.updateVersion) {
                warnings.push(createPeerWarning(`an invalid or missing peer dependency on Appium. A newer version of \`${pkgName}\` is available; ` +
                    `please attempt to upgrade "${extName}" to v${extListData.updateVersion} or newer.`));
            }
            else {
                warnings.push(createPeerWarning(`an invalid or missing peer dependency on Appium. ` +
                    `Please ask the developer of \`${pkgName}\` to add a peer dependency on \`^appium@${build_1.APPIUM_VER}\`.`));
            }
        }
        return warnings;
    }
    /** Validates and registers extension CLI/config schema when the manifest defines a `schema` field. */
    async getSchemaProblems(extManifest, extName) {
        const problems = [];
        const { schema: argSchemaPath } = extManifest;
        if (ExtensionConfig.extDataHasSchema(extManifest)) {
            if (lodash_1.default.isString(argSchemaPath)) {
                if ((0, schema_1.isAllowedSchemaFileExtension)(argSchemaPath)) {
                    try {
                        await this.readExtensionSchema(extName, extManifest);
                    }
                    catch (err) {
                        problems.push({
                            err: `Unable to register schema at path ${argSchemaPath}; ${err.message}`,
                            val: argSchemaPath,
                        });
                    }
                }
                else {
                    problems.push({
                        err: `Schema file has unsupported extension. Allowed: ${[
                            ...schema_1.ALLOWED_SCHEMA_EXTENSIONS,
                        ].join(', ')}`,
                        val: argSchemaPath,
                    });
                }
            }
            else if (lodash_1.default.isPlainObject(argSchemaPath)) {
                try {
                    await this.readExtensionSchema(extName, extManifest);
                }
                catch (err) {
                    problems.push({
                        err: `Unable to register embedded schema; ${err.message}`,
                        val: argSchemaPath,
                    });
                }
            }
            else {
                problems.push({
                    err: 'Incorrectly formatted schema field; must be a path to a schema file or a schema object.',
                    val: argSchemaPath,
                });
            }
        }
        return problems;
    }
    /** Blocking issues for required manifest fields shared by all extensions (version, package name, main class). */
    getGenericConfigProblems(extManifest, extName) {
        void extName;
        const { version, pkgName, mainClass } = extManifest;
        const problems = [];
        if (!lodash_1.default.isString(version)) {
            problems.push({
                err: `Invalid or missing \`version\` field in my \`package.json\` and/or \`extensions.yaml\` (must be a string)`,
                val: version,
            });
        }
        if (!lodash_1.default.isString(pkgName)) {
            problems.push({
                err: `Invalid or missing \`name\` field in my \`package.json\` and/or \`extensions.yaml\` (must be a string)`,
                val: pkgName,
            });
        }
        if (!lodash_1.default.isString(mainClass)) {
            problems.push({
                err: `Invalid or missing \`appium.mainClass\` field in my \`package.json\` and/or \`mainClass\` field in \`extensions.yaml\` (must be a string)`,
                val: mainClass,
            });
        }
        return problems;
    }
    /** Driver- or plugin-specific blocking validation; override in subclasses when needed. */
    getConfigProblems(_extManifest, _extName) {
        void _extManifest;
        void _extName;
        return [];
    }
    async _resolveExtension(extName) {
        const { mainClass } = this.installedExtensions[extName];
        const moduleRoot = this.getInstallPath(extName);
        const packageJsonPath = node_path_1.default.join(moduleRoot, 'package.json');
        let extensionManifest;
        try {
            extensionManifest = JSON.parse(await support_1.fs.readFile(packageJsonPath, 'utf8'));
        }
        catch (e) {
            throw new ReferenceError(`Could not read the ${this.extensionType} manifest at ${packageJsonPath}: ${e.message}`, { cause: e });
        }
        let entryPointRelativePath;
        try {
            if (extensionManifest.type === 'module' && extensionManifest.exports) {
                entryPointRelativePath = resolveEsmEntryPoint(extensionManifest.exports);
            }
            entryPointRelativePath = entryPointRelativePath ?? extensionManifest.main ?? DEFAULT_ENTRY_POINT;
        }
        catch (e) {
            throw new ReferenceError(`Could not find the ${this.extensionType} installed at ${moduleRoot}: ${e.message}`, { cause: e });
        }
        const entryPointFullPath = node_path_1.default.resolve(moduleRoot, entryPointRelativePath);
        if (!await support_1.fs.exists(entryPointFullPath)) {
            throw new ReferenceError(`Cannot find a valid ${this.extensionType} main entry point in '${packageJsonPath}'. ` +
                `Assumed entry point: '${entryPointFullPath}'`);
        }
        // note: this will only reload the entry point
        if (process.env.APPIUM_RELOAD_EXTENSIONS && require.cache[entryPointFullPath]) {
            logger_1.log.debug(`Removing ${entryPointFullPath} from require cache`);
            delete require.cache[entryPointFullPath];
        }
        return [entryPointFullPath, mainClass];
    }
}
exports.ExtensionConfig = ExtensionConfig;
/**
 * Resolves a package `exports` field (string, `"."`, or `"import"`) to a relative entry path for ESM packages.
 *
 * @param exportsValue - `package.json` `exports` value or nested fragment
 */
function resolveEsmEntryPoint(exportsValue) {
    if (lodash_1.default.isString(exportsValue) && exportsValue) {
        return exportsValue;
    }
    if (!lodash_1.default.isPlainObject(exportsValue)) {
        return;
    }
    const obj = exportsValue;
    for (const key of ['.', 'import']) {
        if (obj[key]) {
            return resolveEsmEntryPoint(obj[key]);
        }
    }
}
