// appium/build/lib/extension/manifest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Manifest = void 0;
const bluebird_1 = __importDefault(require("bluebird"));
const support_1 = require("@appium/support");
const lodash_1 = __importDefault(require("lodash"));
const node_path_1 = __importDefault(require("node:path"));
const YAML = __importStar(require("yaml"));
const constants_1 = require("../constants");
const extension_config_1 = require("./extension-config");
const package_changed_1 = require("./package-changed");
const manifest_migrations_1 = require("./manifest-migrations");
const CONFIG_DATA_DRIVER_KEY = `${constants_1.DRIVER_TYPE}s`;
const CONFIG_DATA_PLUGIN_KEY = `${constants_1.PLUGIN_TYPE}s`;
const INITIAL_MANIFEST_DATA = Object.freeze({
    [CONFIG_DATA_DRIVER_KEY]: Object.freeze({}),
    [CONFIG_DATA_PLUGIN_KEY]: Object.freeze({}),
    schemaRev: constants_1.CURRENT_SCHEMA_REV,
});
/**
 * Handles reading & writing of extension config files.
 *
 * Only one instance of this class exists per value of `APPIUM_HOME`.
 */
class Manifest {
    /**
     * Returns the memoized manifest for an `APPIUM_HOME` directory (one instance per home).
     *
     * @param appiumHome - `APPIUM_HOME` path used as the cache key
     */
    static getInstance = lodash_1.default.memoize((appiumHome) => new Manifest(appiumHome));
    #data;
    #appiumHome;
    #manifestPath = undefined;
    #writing;
    #reading;
    constructor(appiumHome) {
        this.#appiumHome = appiumHome;
        this.#data = lodash_1.default.cloneDeep(INITIAL_MANIFEST_DATA);
    }
    /** `APPIUM_HOME` directory this manifest is tied to. */
    get appiumHome() {
        return this.#appiumHome;
    }
    /**
     * Absolute path to the extension manifest file after {@link Manifest.read} or {@link Manifest.write} has resolved it.
     * Before that, this is `undefined`.
     */
    get manifestPath() {
        return this.#manifestPath;
    }
    /** Schema revision of the in-memory manifest data (from YAML `schemaRev`). */
    get schemaRev() {
        return this.#data.schemaRev;
    }
    /**
     * Returns the live installed-extension map for drivers or plugins (same object as stored in memory).
     * Mutations affect the manifest until replaced by a new object (e.g. via `read()`); `setExtension` / `deleteExtension` update this record.
     *
     * @param extType - `"driver"` or `"plugin"`
     */
    getExtensionData(extType) {
        const record = extType === constants_1.DRIVER_TYPE ? this.#data.drivers : this.#data.plugins;
        return record;
    }
    /**
     * Whether a driver with the given manifest key is present.
     *
     * @param name - Driver name as stored under `drivers` in the manifest
     */
    hasDriver(name) {
        return Boolean(this.#data.drivers[name]);
    }
    /**
     * Whether a plugin with the given manifest key is present.
     *
     * @param name - Plugin name as stored under `plugins` in the manifest
     */
    hasPlugin(name) {
        return Boolean(this.#data.plugins[name]);
    }
    /**
     * Loads manifest YAML from disk into memory, runs migration when needed, may sync with installed packages, and writes back if required.
     * Concurrent calls while a read is in flight share the same in-flight work.
     *
     * @returns The parsed in-memory manifest data
     */
    async read() {
        if (this.#reading) {
            await this.#reading;
            return this.#data;
        }
        this.#reading = (async () => {
            let data;
            let shouldWrite = false;
            const manifestPathResolved = await this.#setManifestPath();
            try {
                const yaml = await support_1.fs.readFile(manifestPathResolved, 'utf8');
                data = YAML.parse(yaml);
            }
            catch (err) {
                if (err.code === 'ENOENT') {
                    data = lodash_1.default.cloneDeep(INITIAL_MANIFEST_DATA);
                    shouldWrite = true;
                }
                else {
                    throw new Error(`Appium had trouble loading the extension installation ` +
                        `cache file (${manifestPathResolved}). It may be invalid YAML. Specific error: ${err.message}`, { cause: err });
                }
            }
            this.#data = data;
            if (!shouldWrite && (data.schemaRev ?? 0) < constants_1.CURRENT_SCHEMA_REV) {
                shouldWrite = await (0, manifest_migrations_1.migrate)(this);
            }
            const hasAppiumDependency = await support_1.env.hasAppiumDependency(this.appiumHome);
            if (shouldWrite || (hasAppiumDependency && (await (0, package_changed_1.packageDidChange)(this.appiumHome)))) {
                shouldWrite = (await this.syncWithInstalledExtensions(hasAppiumDependency)) || shouldWrite;
            }
            if (shouldWrite) {
                await this.write();
            }
        })();
        try {
            await this.#reading;
            return this.#data;
        }
        finally {
            this.#reading = undefined;
        }
    }
    /**
     * Serializes the current in-memory manifest to the resolved manifest path.
     * Concurrent calls while a write is in flight share the same in-flight work.
     *
     * @returns `true` when the file was written successfully
     */
    async write() {
        if (this.#writing) {
            return this.#writing;
        }
        this.#writing = (async () => {
            const manifestPathResolved = await this.#setManifestPath();
            try {
                await support_1.fs.mkdirp(node_path_1.default.dirname(manifestPathResolved));
            }
            catch (err) {
                throw new Error(`Appium could not create the directory for the manifest file: ${node_path_1.default.dirname(manifestPathResolved)}. Original error: ${err.message}`, { cause: err });
            }
            try {
                await support_1.fs.writeFile(manifestPathResolved, YAML.stringify(this.#data), 'utf8');
                return true;
            }
            catch (err) {
                throw new Error(`Appium could not write to manifest at ${manifestPathResolved} using APPIUM_HOME ${this.#appiumHome}. Please ensure it is writable. Original error: ${err.message}`, { cause: err });
            }
        })();
        try {
            return await this.#writing;
        }
        finally {
            this.#writing = undefined;
        }
    }
    /**
     * Scans `APPIUM_HOME` (root and `node_modules`) for Appium extension packages and merges them into the manifest.
     *
     * @param hasAppiumDependency - When true and the root `package.json` depends on Appium, matching extensions use the `"dev"` install type
     * @returns `true` if any extension entries changed, `false` otherwise
     */
    async syncWithInstalledExtensions(hasAppiumDependency = false) {
        let didChange = false;
        const onMatch = async (filepath, devType = false) => {
            try {
                const pkg = JSON.parse(await support_1.fs.readFile(filepath, 'utf8'));
                if (isExtension(pkg)) {
                    const installType = devType && hasAppiumDependency ? extension_config_1.INSTALL_TYPE_DEV : extension_config_1.INSTALL_TYPE_NPM;
                    const changed = this.addExtensionFromPackage(pkg, filepath, installType);
                    didChange = didChange || changed;
                }
            }
            catch {
                // ignore invalid package.json
            }
        };
        const queue = [
            onMatch(node_path_1.default.join(this.#appiumHome, 'package.json'), true),
        ];
        const filepaths = await support_1.fs.glob('node_modules/{*,@*/*}/package.json', {
            cwd: this.#appiumHome,
            absolute: true,
        });
        for (const filepath of filepaths) {
            queue.push(onMatch(filepath));
        }
        await bluebird_1.default.all(queue);
        return didChange;
    }
    /**
     * Builds manifest metadata from a `package.json` and registers it if it is a driver or plugin and the entry changed.
     *
     * @param pkgJson - Parsed extension `package.json`
     * @param pkgPath - Path to that `package.json` (install path is derived from its directory)
     * @param installType - How the package was discovered (`npm` vs `dev`)
     * @returns `true` if the manifest was updated, `false` if unchanged or already matched
     * @throws TypeError if the package is not a valid driver or plugin extension
     */
    addExtensionFromPackage(pkgJson, pkgPath, installType = extension_config_1.INSTALL_TYPE_NPM) {
        const extensionPath = node_path_1.default.dirname(pkgPath);
        const internal = {
            pkgName: pkgJson.name,
            version: pkgJson.version,
            appiumVersion: pkgJson.peerDependencies?.appium,
            installType,
            installSpec: `${pkgJson.name}@${pkgJson.version}`,
            installPath: extensionPath,
        };
        if (isDriver(pkgJson)) {
            const driverName = pkgJson.appium.driverName;
            const value = {
                ...lodash_1.default.omit(pkgJson.appium, 'driverName'),
                ...internal,
            };
            if (!lodash_1.default.isEqual(value, this.#data.drivers[driverName])) {
                this.setExtension(constants_1.DRIVER_TYPE, driverName, value);
                return true;
            }
            return false;
        }
        if (isPlugin(pkgJson)) {
            const pluginName = pkgJson.appium.pluginName;
            const value = {
                ...lodash_1.default.omit(pkgJson.appium, 'pluginName'),
                ...internal,
            };
            if (!lodash_1.default.isEqual(value, this.#data.plugins[pluginName])) {
                this.setExtension(constants_1.PLUGIN_TYPE, pluginName, value);
                return true;
            }
            return false;
        }
        throw new TypeError(`The extension in ${extensionPath} is neither a valid ${constants_1.DRIVER_TYPE} nor a valid ${constants_1.PLUGIN_TYPE}.`);
    }
    /**
     * Stores a deep-cloned copy of extension metadata under the given type and name.
     *
     * @param extType - `"driver"` or `"plugin"`
     * @param extName - Manifest key for the extension
     * @param extData - Full extension entry to persist in memory
     * @returns The cloned data now held in the manifest
     */
    setExtension(extType, extName, extData) {
        const data = lodash_1.default.cloneDeep(extData);
        if (extType === constants_1.DRIVER_TYPE) {
            this.#data.drivers[extName] = data;
        }
        else {
            this.#data.plugins[extName] = data;
        }
        return data;
    }
    /**
     * Updates the in-memory manifest schema revision (typically during migration).
     *
     * @param rev - New `schemaRev` value
     */
    setSchemaRev(rev) {
        this.#data.schemaRev = rev;
    }
    /**
     * Removes an extension entry from the manifest data in memory (does not write to disk by itself).
     *
     * @param extType - `"driver"` or `"plugin"`
     * @param extName - Manifest key to remove
     */
    deleteExtension(extType, extName) {
        if (extType === constants_1.DRIVER_TYPE) {
            delete this.#data.drivers[extName];
        }
        else {
            delete this.#data.plugins[extName];
        }
    }
    async #setManifestPath() {
        if (!this.#manifestPath) {
            const resolved = await support_1.env.resolveManifestPath(this.#appiumHome);
            this.#manifestPath = resolved;
            if (node_path_1.default.relative(this.#appiumHome, resolved).startsWith('.')) {
                throw new Error(`Mismatch between location of APPIUM_HOME and manifest file. APPIUM_HOME: ${this.appiumHome}, manifest file: ${resolved}`);
            }
        }
        return this.#manifestPath;
    }
}
exports.Manifest = Manifest;
function isExtension(value) {
    return (lodash_1.default.isPlainObject(value) &&
        lodash_1.default.isPlainObject(value.appium) &&
        lodash_1.default.isString(value.name) &&
        lodash_1.default.isString(value.version));
}
function isDriver(value) {
    return (isExtension(value) &&
        'driverName' in value.appium &&
        lodash_1.default.isString(value.appium.driverName));
}
function isPlugin(value) {
    return (isExtension(value) &&
        'pluginName' in value.appium &&
        lodash_1.default.isString(value.appium.pluginName));
}
