// appium/build/lib/extension/plugin-config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginConfig = void 0;
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const logger_1 = require("../logger");
const extension_config_1 = require("./extension-config");
class PluginConfig extends extension_config_1.ExtensionConfig {
    static _instances = new WeakMap();
    constructor(manifest) {
        super(constants_1.PLUGIN_TYPE, manifest);
    }
    static create(manifest) {
        const instance = new PluginConfig(manifest);
        if (PluginConfig.getInstance(manifest)) {
            throw new Error(`Manifest with APPIUM_HOME ${manifest.appiumHome} already has a PluginConfig; use PluginConfig.getInstance() to retrieve it.`);
        }
        PluginConfig._instances.set(manifest, instance);
        return instance;
    }
    static getInstance(manifest) {
        return PluginConfig._instances.get(manifest);
    }
    async validate() {
        return await super._validate(this.manifest.getExtensionData(constants_1.PLUGIN_TYPE));
    }
    extensionDesc(pluginName, { version }) {
        return `${String(pluginName)}@${version}`;
    }
    print(activeNames = []) {
        const pluginNames = Object.keys(this.installedExtensions);
        if (lodash_1.default.isEmpty(pluginNames)) {
            logger_1.log.info(`No plugins have been installed. Use the "appium plugin" ` + 'command to install the one(s) you want to use.');
            return;
        }
        logger_1.log.info(`Available plugins:`);
        for (const [pluginName, pluginData] of lodash_1.default.toPairs(this.installedExtensions)) {
            const activeTxt = lodash_1.default.includes(activeNames, pluginName) ? ' (ACTIVE)' : '';
            logger_1.log.info(`  - ${this.extensionDesc(pluginName, pluginData)}${activeTxt}`);
        }
        if (lodash_1.default.isEmpty(activeNames)) {
            logger_1.log.info('No plugins activated. Use the --use-plugins flag with names of plugins to activate');
        }
    }
}
exports.PluginConfig = PluginConfig;
