// appium/build/lib/bootstrap/appium-main-runner.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppiumMainRunner = void 0;
const logger_1 = require("../logger");
const lodash_1 = __importDefault(require("lodash"));
const extension_1 = require("../extension");
const grid_v3_register_1 = __importDefault(require("./grid-v3-register"));
const main_helpers_1 = require("./main-helpers");
const node_net_1 = __importDefault(require("node:net"));
const MAX_SERVER_PROCESS_LISTENERS = 100;
/**
 * Starts the Appium HTTP server after {@link AppiumInitializer.init}: loads drivers/plugins, binds, grid register, signals.
 */
class AppiumMainRunner {
    /**
     * For server init: builds listeners, registers with Grid 3 if configured, and returns the server.
     * For non-server commands, `initResult` is empty and this resolves to `undefined`.
     *
     * @param initResult - Output of {@link AppiumInitializer.init}
     * @param args - Original args (used to describe `appiumHome` source in logs)
     */
    async run(initResult, args) {
        if (lodash_1.default.isEmpty(initResult)) {
            return undefined;
        }
        const { appiumDriver, pluginConfig, driverConfig, parsedArgs, appiumHome } = initResult;
        const pluginClasses = await (0, extension_1.getActivePlugins)(pluginConfig, parsedArgs.pluginsImportChunkSize, parsedArgs.usePlugins);
        for (const [pluginClass, name] of pluginClasses) {
            appiumDriver.pluginClasses.set(pluginClass, name);
        }
        await (0, main_helpers_1.logStartupInfo)(parsedArgs);
        appiumDriver.configureGlobalFeatures();
        const appiumHomeSourceName = (0, main_helpers_1.determineAppiumHomeSource)(args?.appiumHome);
        logger_1.log.debug(`The ${appiumHomeSourceName}: ${appiumHome}`);
        const driverClasses = await (0, extension_1.getActiveDrivers)(driverConfig, parsedArgs.driversImportChunkSize, parsedArgs.useDrivers);
        const { serverOpts, normalizedBasePath } = (0, main_helpers_1.buildServerOpts)(appiumDriver, parsedArgs, driverClasses, pluginClasses);
        const server = await this.startHttpServer(serverOpts, appiumDriver, normalizedBasePath);
        if (!server) {
            return undefined;
        }
        this.warnIfCorsEnabled(parsedArgs);
        appiumDriver.server = server;
        await this.registerGridOrClose(server, parsedArgs, normalizedBasePath);
        this.attachSignalHandlers(appiumDriver, server);
        this.logListeningUrl(server, parsedArgs, normalizedBasePath);
        driverConfig.print();
        pluginConfig.print([...pluginClasses.values()]);
        return server;
    }
    async startHttpServer(serverOpts, appiumDriver, normalizedBasePath) {
        try {
            return await (0, main_helpers_1.createAppiumServer)(serverOpts, appiumDriver, normalizedBasePath);
        }
        catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            const stack = err instanceof Error ? err.stack : undefined;
            logger_1.log.error(`Could not configure Appium server. It's possible that a driver or plugin tried ` +
                `to update the server and failed. Original error: ${message}`);
            logger_1.log.debug(stack);
            process.exit(1);
            return undefined;
        }
    }
    warnIfCorsEnabled(parsedArgs) {
        if (parsedArgs.allowCors) {
            logger_1.log.warn('You have enabled CORS requests from any host. Be careful not ' +
                'to visit sites which could maliciously try to start Appium ' +
                'sessions on your machine');
        }
    }
    async registerGridOrClose(server, parsedArgs, normalizedBasePath) {
        try {
            if (parsedArgs.nodeconfig) {
                await (0, grid_v3_register_1.default)(parsedArgs.nodeconfig, parsedArgs.address, parsedArgs.port, normalizedBasePath);
            }
        }
        catch (err) {
            await server.close();
            throw err;
        }
    }
    attachSignalHandlers(appiumDriver, server) {
        process.setMaxListeners(MAX_SERVER_PROCESS_LISTENERS);
        for (const signal of ['SIGINT', 'SIGTERM']) {
            process.once(signal, async function onSignal() {
                logger_1.log.info(`Received ${signal} - shutting down`);
                try {
                    await appiumDriver.shutdown(`The process has received ${signal} signal`);
                    await server.close();
                    process.exit(0);
                }
                catch (e) {
                    logger_1.log.warn(e);
                    process.exit(1);
                }
            });
        }
    }
    logListeningUrl(server, parsedArgs, normalizedBasePath) {
        const protocol = server.isSecure() ? 'https' : 'http';
        const address = node_net_1.default.isIPv6(parsedArgs.address) ? `[${parsedArgs.address}]` : parsedArgs.address;
        (0, main_helpers_1.logServerAddress)(`${protocol}://${address}:${parsedArgs.port}${normalizedBasePath}`);
    }
}
exports.AppiumMainRunner = AppiumMainRunner;
