// appium/build/lib/bootstrap/appium-initializer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppiumInitializer = void 0;
const logsink_1 = require("../logsink");
const logger_1 = require("../logger");
const support_1 = require("@appium/support");
const lodash_1 = __importDefault(require("lodash"));
const appium_1 = require("../appium");
const extension_1 = require("../cli/extension");
const setup_command_1 = require("../cli/setup-command");
const parser_1 = require("../cli/parser");
const config_file_1 = require("./config-file");
const startup_config_1 = require("./startup-config");
const node_helpers_1 = require("./node-helpers");
const extension_2 = require("../extension");
const constants_1 = require("../constants");
const extension_command_1 = require("../cli/extension-command");
const schema_1 = require("../schema/schema");
const cli_args_guards_1 = require("../schema/cli-args-guards");
const main_helpers_1 = require("./main-helpers");
/**
 * Parses CLI/programmatic args, loads config and extensions, and returns server-ready state or runs extension/setup flows.
 */
class AppiumInitializer {
    throwInsteadOfExit = false;
    /**
     * Resolves Appium home, loads config, and either returns server-ready state for the server subcommand
     * or performs setup/extension CLI work and returns an empty result.
     *
     * @param args - Optional programmatic args; when omitted, parses `process.argv`
     */
    async init(args) {
        this.throwInsteadOfExit = false;
        const appiumHome = args?.appiumHome ?? (await support_1.env.resolveAppiumHome());
        const appiumHomeSourceName = (0, main_helpers_1.determineAppiumHomeSource)(args?.appiumHome);
        await (0, node_helpers_1.requireDir)(appiumHome, false, appiumHomeSourceName);
        (0, node_helpers_1.adjustNodePath)();
        const { driverConfig, pluginConfig } = await (0, extension_2.loadExtensions)(appiumHome);
        const { preConfigArgs, throwInsteadOfExit } = await this.parsePreConfigArgs(args);
        this.throwInsteadOfExit = throwInsteadOfExit;
        const configResult = await (0, config_file_1.readConfigFile)(preConfigArgs.configFile);
        this.assertConfigFileOk(configResult);
        if ((0, cli_args_guards_1.isServerCommandArgs)(preConfigArgs)) {
            return this.finishServerInit(preConfigArgs, configResult, driverConfig, pluginConfig, appiumHome, appiumHomeSourceName);
        }
        if ((0, cli_args_guards_1.isSetupCommandArgs)(preConfigArgs)) {
            await (0, setup_command_1.runSetupCommand)(preConfigArgs, driverConfig, pluginConfig);
            return {};
        }
        await (0, node_helpers_1.requireDir)(appiumHome, true, appiumHomeSourceName);
        await this.runExtensionCliIfNeeded(preConfigArgs, driverConfig, pluginConfig);
        return {};
    }
    async parsePreConfigArgs(args) {
        const parser = await (0, parser_1.getParser)();
        let throwInsteadOfExit = false;
        if (args) {
            if (args.throwInsteadOfExit) {
                throwInsteadOfExit = true;
                delete args.throwInsteadOfExit;
            }
            const preConfigArgs = { ...args, subcommand: args.subcommand ?? constants_1.SERVER_SUBCOMMAND };
            parser_1.ArgParser.normalizeServerArgs(preConfigArgs);
            return { preConfigArgs, throwInsteadOfExit };
        }
        return { preConfigArgs: parser.parseArgs(), throwInsteadOfExit };
    }
    assertConfigFileOk(configResult) {
        if (!lodash_1.default.isEmpty(configResult.errors)) {
            throw new Error(`Errors in config file ${configResult.filepath}:\n ${configResult.reason ?? configResult.errors}`);
        }
    }
    async finishServerInit(preConfigArgs, configResult, driverConfig, pluginConfig, appiumHome, appiumHomeSourceName) {
        const defaults = (0, schema_1.getDefaultsForSchema)(false);
        const serverArgs = lodash_1.default.defaultsDeep({}, preConfigArgs, configResult.config?.server, defaults);
        if (preConfigArgs.showConfig) {
            (0, startup_config_1.showConfig)((0, startup_config_1.getNonDefaultServerArgs)(preConfigArgs), configResult, defaults, serverArgs);
            return {};
        }
        if (preConfigArgs.showDebugInfo) {
            await (0, node_helpers_1.showDebugInfo)({
                driverConfig,
                pluginConfig,
                appiumHome,
            });
            return {};
        }
        await (0, logsink_1.init)(serverArgs);
        await this.applyLogFilters(serverArgs);
        if (!serverArgs.noPermsCheck) {
            await (0, node_helpers_1.requireDir)(appiumHome, true, appiumHomeSourceName);
        }
        const appiumDriver = new appium_1.AppiumDriver(serverArgs);
        appiumDriver.driverConfig = driverConfig;
        await (0, main_helpers_1.preflightChecks)(serverArgs, this.throwInsteadOfExit);
        return {
            appiumDriver,
            parsedArgs: serverArgs,
            driverConfig,
            pluginConfig,
            appiumHome,
        };
    }
    async applyLogFilters(serverArgs) {
        if (!serverArgs.logFilters) {
            return;
        }
        const { issues, rules } = await logger_1.log.unwrap().loadSecureValuesPreprocessingRules(serverArgs.logFilters);
        const argToLog = lodash_1.default.truncate(JSON.stringify(serverArgs.logFilters), {
            length: 150,
        });
        if (!lodash_1.default.isEmpty(issues)) {
            throw new Error(`The log filtering rules config ${argToLog} has issues: ` + JSON.stringify(issues, null, 2));
        }
        if (lodash_1.default.isEmpty(rules)) {
            logger_1.log.warn(`Found no log filtering rules in the ${argToLog} config. ` + `Is that expected?`);
        }
        else {
            logger_1.log.info(`Loaded ${support_1.util.pluralize('filtering rule', rules.length, true)}`);
        }
    }
    async runExtensionCliIfNeeded(preConfigArgs, driverConfig, pluginConfig) {
        if (!(0, cli_args_guards_1.isExtensionCommandArgs)(preConfigArgs)) {
            return;
        }
        if ((0, cli_args_guards_1.isDriverCommandArgs)(preConfigArgs)) {
            await (0, extension_1.runExtensionCommand)(preConfigArgs, driverConfig);
        }
        if ((0, cli_args_guards_1.isPluginCommandArgs)(preConfigArgs)) {
            await (0, extension_1.runExtensionCommand)(preConfigArgs, pluginConfig);
        }
        if ((0, cli_args_guards_1.isDriverCommandArgs)(preConfigArgs) || (0, cli_args_guards_1.isPluginCommandArgs)(preConfigArgs)) {
            const cmd = (0, cli_args_guards_1.isDriverCommandArgs)(preConfigArgs)
                ? preConfigArgs.driverCommand
                : preConfigArgs.pluginCommand;
            if (cmd === 'install') {
                await (0, extension_command_1.injectAppiumSymlinks)(driverConfig, pluginConfig, logger_1.log);
            }
        }
    }
}
exports.AppiumInitializer = AppiumInitializer;
