// appium/build/lib/bootstrap/config-file.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readConfigFile = readConfigFile;
exports.normalizeConfig = normalizeConfig;
const lilconfig_1 = require("lilconfig");
const lodash_1 = __importDefault(require("lodash"));
const yaml = __importStar(require("yaml"));
const schema_1 = require("../schema/schema");
const format_errors_1 = require("../schema/format-errors");
/**
 * A cache of the raw config file (a JSON string) at a filepath.
 * This is used for better error reporting.
 * Note that config files needn't be JSON, but it helps if they are.
 */
const rawConfig = new Map();
/**
 * Given an optional path, read a config file. Validates the config file.
 *
 * Call {@link validate} if you already have a config object.
 * @public
 * @param filepath - Explicit config path; when omitted, searches with lilconfig
 * @param opts - e.g. `pretty` for formatted validation errors
 * @returns Contains config and filepath, if found, and any errors
 */
async function readConfigFile(filepath, opts = {}) {
    const lc = (0, lilconfig_1.lilconfig)('appium', {
        loaders: {
            '.yaml': yamlLoader,
            '.yml': yamlLoader,
            '.json': jsonLoader,
            noExt: jsonLoader,
        },
        packageProp: 'appiumConfig',
    });
    const result = filepath ? await loadConfigFile(lc, filepath) : await searchConfigFile(lc);
    if (result?.filepath && !result?.isEmpty) {
        const { pretty = true } = opts;
        try {
            let configResult;
            const errors = (0, schema_1.validate)(result.config);
            if (lodash_1.default.isEmpty(errors)) {
                configResult = { ...result, errors };
            }
            else {
                const reason = (0, format_errors_1.formatErrors)(errors, result.config, {
                    json: rawConfig.get(result.filepath),
                    pretty,
                });
                configResult = reason ? { ...result, errors, reason } : { ...result, errors };
            }
            // normalize (to camel case) all top-level property names of the config file
            configResult.config = normalizeConfig(configResult.config);
            return configResult;
        }
        finally {
            // clean up the raw config file cache, which is only kept to better report errors.
            rawConfig.delete(result.filepath);
        }
    }
    return result ?? {};
}
/**
 * Converts schema property names to either the `appiumCliDest` value, if any, or camelCase.
 *
 * @param config - Raw config object as parsed from file
 * @returns New object with camel-cased keys (or CLI `dest` keys).
 */
function normalizeConfig(config) {
    const schema = (0, schema_1.getSchema)();
    const isSchemaTypeObject = (schemaObj) => Boolean(schemaObj?.properties || schemaObj?.type === 'object');
    const normalize = (rootConfig, section) => {
        const obj = lodash_1.default.isUndefined(section)
            ? rootConfig
            : lodash_1.default.get(rootConfig, section, rootConfig);
        const mappedObj = lodash_1.default.mapKeys(obj, (_v, prop) => lodash_1.default.get(schema, `properties.server.properties[${prop}].appiumCliDest`, lodash_1.default.camelCase(prop)));
        return lodash_1.default.mapValues(mappedObj, (value, property) => {
            const nextSection = section ? `${section}.${property}` : property;
            return isSchemaTypeObject(schema.properties?.[property])
                ? normalize(rootConfig, nextSection)
                : value;
        });
    };
    return normalize(config);
}
/**
 * lilconfig loader to handle `.yaml` files
 */
function yamlLoader(filepath, content) {
    try {
        return yaml.parse(content);
    }
    catch (e) {
        throw new Error(`The YAML config at '${filepath}' cannot be loaded. Original error: ${e.message}`, { cause: e });
    }
}
/**
 * Custom JSON loader that caches the raw config file (for use with `better-ajv-errors`).
 * If it weren't for this cache, this would be unnecessary.
 */
function jsonLoader(filepath, content) {
    rawConfig.set(filepath, content);
    try {
        return JSON.parse(content);
    }
    catch (e) {
        rawConfig.delete(filepath);
        throw new Error(`The JSON config at '${filepath}' cannot be loaded. Original error: ${e.message}`, { cause: e });
    }
}
/**
 * Loads a config file from an explicit path
 */
async function loadConfigFile(lc, filepath) {
    try {
        // removing "await" will cause any rejection to _not_ be caught in this block!
        return await lc.load(filepath);
    }
    catch (err) {
        if (err.code === 'ENOENT') {
            err.message = `Config file not found at user-provided path: ${filepath}`;
            throw err;
        }
        else if (err instanceof SyntaxError) {
            // generally invalid JSON
            err.message = `Config file at user-provided path ${filepath} is invalid:\n${err.message}`;
            throw err;
        }
        throw err;
    }
}
/**
 * Searches for a config file
 */
async function searchConfigFile(lc) {
    return await lc.search();
}
