// appium/build/lib/bootstrap/grid-v3-register.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = registerNode;
const axios_1 = __importDefault(require("axios"));
const support_1 = require("@appium/support");
const lodash_1 = __importDefault(require("lodash"));
const logger_1 = require("../logger");
/**
 * Selenium **Grid 3** (legacy hub) node integration.
 *
 * Registers this Appium process as a node with a Selenium Grid 3 hub using the classic
 * `/grid/register` endpoint and `/grid/api/proxy` for re-registration checks.
 * This does **not** apply to Selenium Grid 4 (which uses a different topology and APIs).
 *
 * @see https://www.selenium.dev/documentation/legacy/grid_3/
 */
/** REST paths exposed by a Selenium Grid 3 hub. */
const GRID_V3_REGISTER_PATH = '/grid/register';
const GRID_V3_PROXY_API_PATH = '/grid/api/proxy';
/**
 * Registers this server as a node with a **Selenium Grid 3** hub.
 *
 * When `data` is a path to a JSON config file, `addr`, `port`, and `basePath` are required (they define
 * how the hub reaches this Appium server). When `data` is an inline config object, those parameters are
 * optional and only used to fill missing node URL fields in the config.
 *
 * @deprecated Selenium Grid 3 registration is legacy and slated for removal.
 * Use Selenium Grid 4 integration (node relay) instead.
 *
 * @param data - Path to a JSON file or an inline parsed config object
 * @param addr - Bind to this address
 * @param port - Bind to this port
 * @param basePath - Base path for the Appium server (used in the node URL sent to the hub; may be `''`)
 */
async function registerNode(data, addr, port, basePath) {
    if (typeof data === 'string') {
        if (addr === undefined || port === undefined || basePath === undefined) {
            throw logger_1.log.errorWithException('When the first argument is a Selenium Grid 3 node config file path, address, port, and basePath ' +
                'are required (e.g. match your Appium `--address`, `--port`, and base path).');
        }
        if (typeof port !== 'number' || !Number.isFinite(port)) {
            throw logger_1.log.errorWithException('When registering from a node config file path, port must be a finite number.');
        }
    }
    let configHolder;
    if (lodash_1.default.isString(data)) {
        const configFilePath = data;
        let fileContent;
        try {
            fileContent = await support_1.fs.readFile(data, 'utf-8');
        }
        catch (err) {
            logger_1.log.error(`Unable to load Selenium Grid 3 node configuration file ${configFilePath} to ` +
                `register with the hub: ${err.message}`);
            return;
        }
        try {
            configHolder = JSON.parse(fileContent);
        }
        catch (err) {
            throw logger_1.log.errorWithException(`Syntax error in Selenium Grid 3 node configuration file ${configFilePath}: ` +
                err.message);
        }
    }
    else {
        configHolder = data;
    }
    postRequest(configHolder, addr, port, basePath);
}
/** Base URL for the Selenium Grid 3 hub (protocol + host + port). */
function hubUri(config) {
    const protocol = config.hubProtocol || 'http';
    return `${protocol}://${config.hubHost}:${config.hubPort}`;
}
/** POST registration payload to the Selenium Grid 3 hub. */
async function registerToGrid(postOptions, configHolder) {
    const hubCfg = configHolder.configuration;
    if (!hubCfg) {
        return;
    }
    try {
        const { status } = await (0, axios_1.default)(postOptions);
        if (status !== 200) {
            throw new Error(`Request failed with code ${status}`);
        }
        logger_1.log.debug(`Appium successfully registered with the Selenium Grid 3 hub at ` + hubUri(hubCfg));
    }
    catch (err) {
        logger_1.log.error(`An attempt to register with the Selenium Grid 3 hub was unsuccessful: ` +
            err.message);
    }
}
function postRequest(configHolder, addr, port, basePath) {
    // Move Selenium Grid 3 (flat) configuration properties into `configuration`
    if (!lodash_1.default.has(configHolder, 'configuration')) {
        const configuration = {};
        const holder = configHolder;
        for (const property in holder) {
            if (lodash_1.default.has(holder, property) && property !== 'capabilities') {
                configuration[property] = holder[property];
                delete holder[property];
            }
        }
        holder.configuration = configuration;
    }
    const cfg = configHolder.configuration;
    if (!cfg) {
        return;
    }
    // if the node config does not have the appium/webdriver url, host, and port,
    // automatically add it based on how appium was initialized
    // otherwise, we will take whatever the user setup
    // because we will always set localhost/127.0.0.1. this won't work if your
    // node and hub aren't in the same place
    if (!cfg.url || !cfg.host || !cfg.port) {
        cfg.url = `http://${addr}:${port}${basePath}`;
        cfg.host = addr;
        cfg.port = port;
    }
    // if the node config does not have id automatically add it
    if (!cfg.id) {
        cfg.id = `http://${cfg.host}:${cfg.port}`;
    }
    // the post options
    const regRequest = {
        url: `${hubUri(cfg)}${GRID_V3_REGISTER_PATH}`,
        method: 'POST',
        data: configHolder,
    };
    if (cfg.register !== true) {
        logger_1.log.debug(`No Selenium Grid 3 hub registration sent (${cfg.register} = false)`);
        return;
    }
    const registerCycleInterval = cfg.registerCycle;
    if (registerCycleInterval === undefined || isNaN(registerCycleInterval) || registerCycleInterval <= 0) {
        logger_1.log.warn(`'registerCycle' is not a valid positive number. ` +
            `No registration request will be sent to the Selenium Grid 3 hub.`);
        return;
    }
    // initiate a new Thread
    let first = true;
    logger_1.log.debug(`Starting auto-register thread for Selenium Grid 3. ` +
        `Will try to register every ${registerCycleInterval} ms.`);
    setInterval(async function registerRetry() {
        if (first) {
            first = false;
            await registerToGrid(regRequest, configHolder);
        }
        else if (!(await isAlreadyRegistered(configHolder))) {
            // make the http POST to the Selenium Grid 3 hub for registration
            await registerToGrid(regRequest, configHolder);
        }
    }, registerCycleInterval);
}
/** Query the Selenium Grid 3 hub to see if this node id is already registered. */
async function isAlreadyRegistered(configHolder) {
    //check if node is already registered
    const hubCfg = configHolder.configuration;
    if (!hubCfg?.id) {
        return;
    }
    const id = hubCfg.id;
    try {
        const { data, status } = await (0, axios_1.default)({
            url: `${hubUri(hubCfg)}${GRID_V3_PROXY_API_PATH}`,
            params: { id },
            timeout: 10000,
        });
        if (status !== 200) {
            throw new Error(`Request failed with code ${status}`);
        }
        if (!data.success) {
            // if register fail, print the debug msg
            logger_1.log.debug(`Selenium Grid 3 hub registration check: ${data.msg}`);
        }
        return data.success;
    }
    catch (err) {
        logger_1.log.debug(`Selenium Grid 3 hub down or not responding: ${err.message}`);
    }
}
