// appium/build/lib/bootstrap/main-helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.inspect = void 0;
exports.showBuildInfo = showBuildInfo;
exports.determineAppiumHomeSource = determineAppiumHomeSource;
exports.logServerAddress = logServerAddress;
exports.preflightChecks = preflightChecks;
exports.logStartupInfo = logStartupInfo;
exports.getServerUpdaters = getServerUpdaters;
exports.getExtraMethodMap = getExtraMethodMap;
exports.buildServerOpts = buildServerOpts;
exports.createAppiumServer = createAppiumServer;
const node_util_1 = require("node:util");
const logger_1 = require("../logger");
const base_driver_1 = require("@appium/base-driver");
const lodash_1 = __importDefault(require("lodash"));
const ws_1 = require("ws");
const build_1 = require("../helpers/build");
const node_helpers_1 = require("./node-helpers");
const startup_config_1 = require("./startup-config");
const schema_1 = require("../schema/schema");
const network_1 = require("../helpers/network");
const constants_1 = require("../constants");
const isStdoutTTY = process.stdout.isTTY;
/**
 * Logs a value to the console using the info logger (with util.inspect formatting).
 */
exports.inspect = lodash_1.default.flow(lodash_1.default.partialRight(node_util_1.inspect, {
    colors: true,
    depth: null,
    compact: !isStdoutTTY,
}), (...args) => {
    logger_1.log.info(...args);
});
/**
 * Prints the current build info JSON to stdout.
 *
 * This updates build metadata first (using GitHub fallback) and then logs the
 * resulting {@link BuildInfo} object.
 */
async function showBuildInfo() {
    await (0, build_1.updateBuildInfo)(true);
    // eslint-disable-next-line no-console -- CLI output for --build-info
    console.log(JSON.stringify((0, build_1.getBuildInfo)()));
}
/**
 * Human-readable label for where Appium resolved `APPIUM_HOME` from (CLI, env, or autodetect).
 */
function determineAppiumHomeSource(appiumHomeFromArgs) {
    if (!lodash_1.default.isNil(appiumHomeFromArgs)) {
        return 'appiumHome config value';
    }
    if (process.env.APPIUM_HOME) {
        return 'APPIUM_HOME environment variable';
    }
    return 'autodetected Appium home path';
}
/**
 * Logs the REST listener URL; if the bind address is a broadcast address, lists concrete interface URLs.
 */
function logServerAddress(url) {
    const urlObj = new URL(url);
    logger_1.log.info(`Appium REST http interface listener started on ${url}`);
    if (!(0, network_1.isBroadcastIp)(urlObj.hostname)) {
        return;
    }
    const interfaces = (0, network_1.fetchInterfaces)(urlObj.hostname === network_1.V4_BROADCAST_IP ? 4 : 6);
    const toLabel = (iface) => {
        const href = urlObj.href.replace(urlObj.hostname, iface.address);
        return iface.internal ? `${href} (only accessible from the same host)` : href;
    };
    logger_1.log.info(`You can provide the following ${interfaces.length === 1 ? 'URL' : 'URLs'} ` +
        `in your client code to connect to this server:\n` +
        interfaces.map((iface) => `\t${toLabel(iface)}`).join('\n'));
}
/**
 * Validates Node version, optional long stack traces, schema, tmp dir, and `--build-info` early exit.
 *
 * @param args - Parsed server CLI args
 * @param throwInsteadOfExit - When true, rethrows failures instead of calling `process.exit(1)`
 */
async function preflightChecks(args, throwInsteadOfExit = false) {
    try {
        (0, node_helpers_1.checkNodeOk)();
        if (args.longStacktrace) {
            Error.stackTraceLimit = constants_1.LONG_STACKTRACE_LIMIT;
        }
        if (args.showBuildInfo) {
            await showBuildInfo();
            process.exit(0);
        }
        (0, schema_1.validate)(args);
        if (args.tmpDir) {
            await (0, node_helpers_1.requireDir)(args.tmpDir, !args.noPermsCheck, 'tmpDir argument value');
        }
    }
    catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        logger_1.log.error(message.red);
        if (throwInsteadOfExit) {
            throw err;
        }
        process.exit(1);
    }
}
/**
 * Prints welcome line (version + optional git rev), non-default server args, and default capabilities.
 */
async function logStartupInfo(args) {
    let welcome = `Welcome to Appium v${build_1.APPIUM_VER}`;
    const appiumRev = await (0, build_1.getGitRev)();
    if (appiumRev) {
        welcome += ` (REV ${appiumRev})`;
    }
    logger_1.log.info(welcome);
    const showArgs = (0, startup_config_1.getNonDefaultServerArgs)(args);
    if (lodash_1.default.size(showArgs)) {
        logNonDefaultArgsWarning(showArgs);
    }
    if (!lodash_1.default.isEmpty(args.defaultCapabilities)) {
        logDefaultCapabilitiesWarning(args.defaultCapabilities);
    }
}
/**
 * Collects `updateServer` hooks from active driver and plugin classes for HTTP server customization.
 */
function getServerUpdaters(driverClasses, pluginClasses) {
    return lodash_1.default.compact(lodash_1.default.map([...driverClasses.keys(), ...pluginClasses.keys()], 'updateServer'));
}
/**
 * Merges `newMethodMap` contributions from all active drivers and plugins into one method map.
 */
function getExtraMethodMap(driverClasses, pluginClasses) {
    return [...driverClasses.keys(), ...pluginClasses.keys()].reduce((map, klass) => ({
        ...map,
        ...(klass.newMethodMap ?? {}),
    }), {});
}
/**
 * Builds {@link ServerOpts} and normalized base path for the Appium HTTP server from CLI args and extensions.
 */
function buildServerOpts(appiumDriver, parsedArgs, driverClasses, pluginClasses) {
    const routeConfiguringFunction = (0, base_driver_1.routeConfiguringFunction)(appiumDriver);
    const serverOpts = {
        routeConfiguringFunction,
        port: parsedArgs.port,
        hostname: parsedArgs.address,
        allowCors: parsedArgs.allowCors,
        basePath: parsedArgs.basePath,
        serverUpdaters: getServerUpdaters(driverClasses, pluginClasses),
        extraMethodMap: getExtraMethodMap(driverClasses, pluginClasses),
        cliArgs: parsedArgs,
    };
    const normalizedBasePath = (0, base_driver_1.normalizeBasePath)(parsedArgs.basePath);
    for (const timeoutArgName of ['keepAliveTimeout', 'requestTimeout']) {
        if (lodash_1.default.isInteger(parsedArgs[timeoutArgName])) {
            serverOpts[timeoutArgName] = parsedArgs[timeoutArgName] * 1000;
        }
    }
    return { serverOpts, normalizedBasePath };
}
/**
 * Creates the Appium HTTP server and attaches WebSocket handlers for BiDi under the normalized base path.
 */
async function createAppiumServer(serverOpts, appiumDriver, normalizedBasePath) {
    const bidiServer = new ws_1.WebSocketServer({ noServer: true });
    bidiServer.on('connection', appiumDriver.onBidiConnection.bind(appiumDriver));
    bidiServer.on('error', appiumDriver.onBidiServerError.bind(appiumDriver));
    const server = await (0, base_driver_1.server)(serverOpts);
    const bidiBasePath = `${normalizedBasePath}${constants_1.BIDI_BASE_PATH}`;
    await server.addWebSocketHandler(bidiBasePath, bidiServer);
    await server.addWebSocketHandler(`${bidiBasePath}/:sessionId`, bidiServer);
    return server;
}
function logNonDefaultArgsWarning(args) {
    logger_1.log.info('Non-default server args:');
    (0, exports.inspect)(args);
}
function logDefaultCapabilitiesWarning(caps) {
    logger_1.log.info('Default capabilities, which will be added to each request ' +
        'unless overridden by desired capabilities:');
    (0, exports.inspect)(caps);
}
