// appium/build/lib/bootstrap/node-helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkNodeOk = checkNodeOk;
exports.adjustNodePath = adjustNodePath;
exports.showDebugInfo = showDebugInfo;
exports.requireDir = requireDir;
/* eslint-disable no-console */
const lodash_1 = __importDefault(require("lodash"));
const node_path_1 = __importDefault(require("node:path"));
const node_os_1 = __importDefault(require("node:os"));
const semver = __importStar(require("semver"));
const bluebird_1 = __importDefault(require("bluebird"));
const support_1 = require("@appium/support");
const logger_1 = require("../logger");
const utils_1 = require("../utils");
const build_1 = require("../helpers/build");
const MIN_NODE_VERSION = utils_1.npmPackage.engines.node;
/**
 * @throws {Error} If Node version is outside of the supported range
 */
function checkNodeOk() {
    const version = getNodeVersion();
    if (!semver.satisfies(version, MIN_NODE_VERSION)) {
        throw new Error(`Node version must be at least ${MIN_NODE_VERSION}; current is ${version.version}`);
    }
}
/**
 * Adjusts NODE_PATH so CJS drivers/plugins can load peer deps. Does not work with ESM.
 */
function adjustNodePath() {
    let appiumModuleSearchRoot;
    try {
        appiumModuleSearchRoot = node_path_1.default.dirname((0, utils_1.getAppiumModuleRoot)());
    }
    catch (error) {
        logger_1.log.warn(error.message);
        return;
    }
    const refreshRequirePaths = () => {
        try {
            // Private API; see https://gist.github.com/branneman/8048520#7-the-hack
            require('node:module').Module._initPaths();
            return true;
        }
        catch {
            return false;
        }
    };
    if (!process.env.NODE_PATH) {
        process.env.NODE_PATH = appiumModuleSearchRoot;
        if (refreshRequirePaths()) {
            process.env.APPIUM_OMIT_PEER_DEPS = '1';
        }
        else {
            delete process.env.NODE_PATH;
        }
        return;
    }
    const nodePathParts = process.env.NODE_PATH.split(node_path_1.default.delimiter);
    if (nodePathParts.includes(appiumModuleSearchRoot)) {
        process.env.APPIUM_OMIT_PEER_DEPS = '1';
        return;
    }
    nodePathParts.push(appiumModuleSearchRoot);
    process.env.NODE_PATH = nodePathParts.join(node_path_1.default.delimiter);
    if (refreshRequirePaths()) {
        process.env.APPIUM_OMIT_PEER_DEPS = '1';
    }
    else {
        process.env.NODE_PATH = lodash_1.default.without(nodePathParts, appiumModuleSearchRoot).join(node_path_1.default.delimiter);
    }
}
/**
 * Prints JSON debug info (OS, Node/npm, Appium build, installed drivers/plugins) to stdout.
 * Input combines extension config snapshots with the resolved Appium home path.
 */
async function showDebugInfo({ driverConfig, pluginConfig, appiumHome }) {
    const getNpmVersion = async () => {
        const { stdout } = await support_1.npm.exec('--version', [], { cwd: process.cwd() });
        return lodash_1.default.trim(stdout);
    };
    const findNpmLocation = async () => await support_1.fs.which(support_1.system.isWindows() ? 'npm.cmd' : 'npm');
    const [npmVersion, npmLocation] = await bluebird_1.default.all([
        ...[getNpmVersion, findNpmLocation].map((f) => getSafeResult(f, 'unknown')),
        (0, build_1.updateBuildInfo)(),
    ]);
    const debugInfo = {
        os: {
            platform: node_os_1.default.platform(),
            release: node_os_1.default.release(),
            arch: node_os_1.default.arch(),
            homedir: node_os_1.default.homedir(),
            username: node_os_1.default.userInfo().username,
        },
        node: {
            version: process.version,
            arch: process.arch,
            cwd: process.cwd(),
            argv: process.argv,
            env: process.env,
            npm: {
                location: npmLocation,
                version: npmVersion,
            },
        },
        appium: {
            location: build_1.rootDir,
            homedir: appiumHome,
            build: (0, build_1.getBuildInfo)(),
            drivers: driverConfig.installedExtensions,
            plugins: pluginConfig.installedExtensions,
        },
    };
    console.log(JSON.stringify(debugInfo, null, 2));
}
/**
 * Ensures a directory exists and (optionally) is writeable.
 *
 * If the directory does not exist, this attempts to create it recursively.
 *
 * @throws {Error}
 */
async function requireDir(root, requireWriteable = true, displayName = 'folder path') {
    let stat;
    try {
        stat = await support_1.fs.stat(root);
    }
    catch (e) {
        if (e.code === 'ENOENT') {
            try {
                await support_1.fs.mkdir(root, { recursive: true });
                return;
            }
            catch { }
        }
        throw new Error(`The ${displayName} '${root}' must exist and be a valid directory`, { cause: e });
    }
    if (stat && !stat.isDirectory()) {
        throw new Error(`The ${displayName} '${root}' must be a valid directory`);
    }
    if (requireWriteable) {
        try {
            await support_1.fs.access(root, support_1.fs.constants.W_OK);
        }
        catch (e) {
            throw new Error(`The ${displayName} '${root}' must be writeable for the current user account '${node_os_1.default.userInfo().username}'`, { cause: e });
        }
    }
}
function getNodeVersion() {
    return semver.coerce(process.version);
}
/**
 * Calculates the result of the given function and return its value
 * or the default one if there was an exception.
 */
async function getSafeResult(f, defaultValue) {
    try {
        return await f();
    }
    catch {
        return defaultValue;
    }
}
