// appium-adb/build/lib/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultsDeep = exports.pick = exports.zip = exports.intersectionWith = exports.defaults = exports.cloneDeep = exports.memoize = exports.getJavaForOs = exports.getJavaHome = exports.requireSdkRoot = exports.getSdkRootFromEnv = exports.unzipFile = exports.getResourcePath = exports.readPackageManifest = exports.buildInstallArgs = exports.DEFAULT_ADB_EXEC_TIMEOUT = exports.APK_INSTALL_TIMEOUT = exports.APKS_EXTENSION = void 0;
const helpers = __importStar(require("./helpers"));
const lodash = __importStar(require("./lodash"));
const support_1 = require("@appium/support");
exports.APKS_EXTENSION = helpers.APKS_EXTENSION;
exports.APK_INSTALL_TIMEOUT = helpers.APK_INSTALL_TIMEOUT;
exports.DEFAULT_ADB_EXEC_TIMEOUT = helpers.DEFAULT_ADB_EXEC_TIMEOUT;
exports.buildInstallArgs = helpers.buildInstallArgs;
exports.readPackageManifest = helpers.readPackageManifest;
exports.getResourcePath = helpers.getResourcePath;
exports.unzipFile = helpers.unzipFile;
exports.getSdkRootFromEnv = helpers.getSdkRootFromEnv;
exports.requireSdkRoot = helpers.requireSdkRoot;
exports.getJavaHome = helpers.getJavaHome;
exports.getJavaForOs = helpers.getJavaForOs;
exports.memoize = support_1.util.memoize;
exports.cloneDeep = lodash.cloneDeep;
exports.defaults = lodash.defaults;
exports.intersectionWith = lodash.intersectionWith;
exports.zip = lodash.zip;
exports.pick = lodash.pick;
exports.defaultsDeep = lodash.defaultsDeep;
