// appium/build/lib/helpers/build.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rootDir = exports.APPIUM_VER = void 0;
exports.getGitRev = getGitRev;
exports.updateBuildInfo = updateBuildInfo;
exports.getBuildInfo = getBuildInfo;
const lodash_1 = __importDefault(require("lodash"));
const axios_1 = __importDefault(require("axios"));
const teen_process_1 = require("teen_process");
const support_1 = require("@appium/support");
const utils_1 = require("../utils");
exports.APPIUM_VER = utils_1.npmPackage.version;
exports.rootDir = support_1.fs.findRoot(__dirname);
const GIT_BINARY = `git${support_1.system.isWindows() ? '.exe' : ''}`;
const GITHUB_API = 'https://api.github.com/repos/appium/appium';
const getFullGitPath = lodash_1.default.memoize(async function getFullGitPath() {
    try {
        return await support_1.fs.which(GIT_BINARY);
    }
    catch {
        return null;
    }
});
/**
 * Returns the current git commit SHA for this Appium checkout.
 *
 * Attempts to read from local git first; when unavailable and fallback is enabled,
 * queries the GitHub API for the tag matching the current Appium version.
 */
async function getGitRev(useGithubApiFallback = false) {
    const fullGitPath = await getFullGitPath();
    if (fullGitPath) {
        try {
            const { stdout } = await (0, teen_process_1.exec)(fullGitPath, ['rev-parse', 'HEAD'], {
                cwd: __dirname,
            });
            return stdout.trim();
        }
        catch { }
    }
    if (!useGithubApiFallback) {
        return null;
    }
    // If the package folder is not a valid git repository
    // then fetch the corresponding tag info from GitHub
    try {
        return (await axios_1.default.get(`${GITHUB_API}/git/refs/tags/appium@${exports.APPIUM_VER}`, {
            headers: {
                'User-Agent': `Appium ${exports.APPIUM_VER}`,
            },
        })).data?.object?.sha;
    }
    catch { }
    return null;
}
async function getGitTimestamp(commitSha, useGithubApiFallback = false) {
    const fullGitPath = await getFullGitPath();
    if (fullGitPath) {
        try {
            const { stdout } = await (0, teen_process_1.exec)(fullGitPath, ['show', '-s', '--format=%ci', commitSha], {
                cwd: __dirname,
            });
            return stdout.trim();
        }
        catch { }
    }
    if (!useGithubApiFallback) {
        return null;
    }
    try {
        return (await axios_1.default.get(`${GITHUB_API}/git/tags/${commitSha}`, {
            headers: {
                'User-Agent': `Appium ${exports.APPIUM_VER}`,
            },
        })).data?.tagger?.date;
    }
    catch { }
    return null;
}
const BUILD_INFO = {
    version: exports.APPIUM_VER,
};
/**
 * Update mutable build info metadata from local git or GitHub fallback.
 */
async function updateBuildInfo(useGithubApiFallback = false) {
    const sha = await getGitRev(useGithubApiFallback);
    if (!sha) {
        return;
    }
    BUILD_INFO['git-sha'] = sha;
    const buildTimestamp = await getGitTimestamp(sha, useGithubApiFallback);
    if (buildTimestamp) {
        BUILD_INFO.built = buildTimestamp;
    }
}
/**
 * Mutable object containing Appium build information. By default it
 * only contains the Appium version, but is updated with the build timestamp
 * and git commit hash asynchronously as soon as `updateBuildInfo` is called
 * and succeeds.
 */
function getBuildInfo() {
    return BUILD_INFO;
}
