// appium/build/lib/helpers/capability.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeNonW3cCapsError = makeNonW3cCapsError;
exports.parseCapsForInnerDriver = parseCapsForInnerDriver;
exports.insertAppiumPrefixes = insertAppiumPrefixes;
exports.removeAppiumPrefixes = removeAppiumPrefixes;
exports.pullSettings = pullSettings;
const lodash_1 = __importDefault(require("lodash"));
const logger_1 = require("../logger");
const base_driver_1 = require("@appium/base-driver");
const W3C_APPIUM_PREFIX = 'appium';
const STANDARD_CAPS_LOWERCASE = new Set([...base_driver_1.STANDARD_CAPS].map((cap) => cap.toLowerCase()));
/**
 * Creates an error when a session receives non-W3C capabilities.
 */
function makeNonW3cCapsError() {
    return new base_driver_1.errors.SessionNotCreatedError('Session capabilities format must comply to the W3C standard. Make sure your client is up to date. ' +
        'See https://www.w3.org/TR/webdriver/#new-session for more details.');
}
/**
 * Parses W3C capabilities for the inner driver and applies defaults.
 *
 * @returns Parsed caps or an invalid result with an error.
 */
function parseCapsForInnerDriver(w3cCapabilities, constraints = {}, defaultCapabilities = {}) {
    if (!(0, base_driver_1.isW3cCaps)(w3cCapabilities)) {
        return { error: makeNonW3cCapsError() };
    }
    let desiredCaps = {};
    // eslint-disable-next-line prefer-const -- assigned in success path after try
    let processedW3CCapabilities;
    w3cCapabilities = lodash_1.default.cloneDeep(w3cCapabilities);
    defaultCapabilities = lodash_1.default.cloneDeep(defaultCapabilities);
    if (!lodash_1.default.isEmpty(defaultCapabilities)) {
        for (const [defaultCapKey, defaultCapValue] of lodash_1.default.toPairs(defaultCapabilities)) {
            let isCapAlreadySet = false;
            for (const firstMatchEntry of w3cCapabilities.firstMatch ?? []) {
                if (lodash_1.default.isPlainObject(firstMatchEntry) &&
                    lodash_1.default.has(removeAppiumPrefixes(firstMatchEntry), removeAppiumPrefix(defaultCapKey))) {
                    isCapAlreadySet = true;
                    break;
                }
            }
            isCapAlreadySet =
                isCapAlreadySet ||
                    (lodash_1.default.isPlainObject(w3cCapabilities.alwaysMatch) &&
                        lodash_1.default.has(removeAppiumPrefixes(w3cCapabilities.alwaysMatch), removeAppiumPrefix(defaultCapKey)));
            if (isCapAlreadySet) {
                continue;
            }
            if (lodash_1.default.isEmpty(w3cCapabilities.firstMatch)) {
                w3cCapabilities.firstMatch = [{ [defaultCapKey]: defaultCapValue }];
            }
            else {
                w3cCapabilities.firstMatch[0][defaultCapKey] = defaultCapValue;
            }
        }
    }
    try {
        desiredCaps = (0, base_driver_1.processCapabilities)(w3cCapabilities, constraints, true);
    }
    catch (error) {
        logger_1.log.info(`Could not parse W3C capabilities: ${error.message}`);
        return {
            desiredCaps,
            processedW3CCapabilities,
            error: error,
        };
    }
    processedW3CCapabilities = {
        alwaysMatch: { ...insertAppiumPrefixes(desiredCaps) },
        firstMatch: [{}],
    };
    return {
        desiredCaps,
        processedW3CCapabilities,
    };
}
/**
 * Prefixes capability keys with `appium:` where appropriate.
 */
function insertAppiumPrefixes(caps) {
    return lodash_1.default.mapKeys(caps, (_, key) => STANDARD_CAPS_LOWERCASE.has(key.toLowerCase()) || key.includes(':')
        ? key
        : `${W3C_APPIUM_PREFIX}:${key}`);
}
/**
 * Removes `appium:` prefix from capability keys.
 */
function removeAppiumPrefixes(caps) {
    return lodash_1.default.mapKeys(caps, (_, key) => removeAppiumPrefix(key));
}
/**
 * Pulls Appium settings from capabilities (mutates caps). Supports
 * `settings[key]: value` and `settings: { key: value }`.
 *
 * @returns Parsed settings object; empty if none found.
 */
function pullSettings(caps) {
    if (!lodash_1.default.isPlainObject(caps) || lodash_1.default.isEmpty(caps)) {
        return {};
    }
    const result = {};
    const singleSettings = {};
    for (const [key, value] of lodash_1.default.toPairs(caps)) {
        let match;
        if (/^(s|appium:s)ettings$/.test(key) && lodash_1.default.isPlainObject(value)) {
            Object.assign(result, value);
            delete caps[key];
        }
        else if ((match = /^(s|appium:s)ettings\[(\S+)\]$/.exec(key))) {
            singleSettings[match[2]] = value;
            delete caps[key];
        }
    }
    if (!lodash_1.default.isEmpty(singleSettings)) {
        Object.assign(result, singleSettings);
    }
    return result;
}
function removeAppiumPrefix(key) {
    const prefix = `${W3C_APPIUM_PREFIX}:`;
    return lodash_1.default.startsWith(key, prefix) ? key.substring(prefix.length) : key;
}
